<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Advanced motif and sequence manipulation • universalmotif</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Advanced motif and sequence manipulation">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">universalmotif</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AdvancedUsage.html">Advanced motif and sequence manipulation</a>
    </li>
    <li>
      <a href="../articles/Introduction.html">Introduction to the "universalmotif" package</a>
    </li>
    <li>
      <a href="../articles/IntroductionToSequenceMotifs.html">Introduction to sequence motifs</a>
    </li>
    <li>
      <a href="../articles/MotifManipulation.html">Motif import, export, visualization, and manipulation</a>
    </li>
    <li>
      <a href="../articles/SequenceSearches.html">Sequence creation, manipulation, and scanning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bjmt/universalmotif">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Advanced motif and sequence manipulation</h1>
                        <h4 class="author">Benjamin Tremblay</h4>
            <address class="author_afil">
      University of Waterloo, Waterloo, Canada<br><a class="author_email" href="mailto:#"></a><a href="mailto:b2tremblay@uwaterloo.ca">b2tremblay@uwaterloo.ca</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/bjmt/universalmotif/blob/master/R/vignettes/AdvancedUsage.Rmd"><code>R/vignettes/AdvancedUsage.Rmd</code></a></small>
      <div class="hidden name"><code>AdvancedUsage.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Introducing sequence motif types.
    </div>
    
<div id="multi-letter-information" class="section level2">
<h2 class="hasAnchor">
<a href="#multi-letter-information" class="anchor"></a>Multi-letter information</h2>
<p>Though PCM/PPM/PWM/ICM type motifs are still widely used today, a few ‘next generation’ motif formats have been proposed. These wish to add another layer of information to motifs: nucleotide interdependence. To illustrate this, consider the following sequences:</p>
<table class="table">
<caption>(#tab:seqs2) Example sequences.</caption>
<thead><tr class="header">
<th>#</th>
<th>Sequence</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>CAAAACC</td>
</tr>
<tr class="even">
<td>2</td>
<td>CAAAACC</td>
</tr>
<tr class="odd">
<td>3</td>
<td>CAAAACC</td>
</tr>
<tr class="even">
<td>4</td>
<td>CTTTTCC</td>
</tr>
<tr class="odd">
<td>5</td>
<td>CTTTTCC</td>
</tr>
<tr class="even">
<td>6</td>
<td>CTTTTCC</td>
</tr>
</tbody>
</table>
<p>This becomes the following PPM:</p>
<table class="table">
<caption>(#tab:ppm2) Position probability matrix.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>0.0</td>
<td>0.5</td>
<td>0.5</td>
<td>0.5</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>C</td>
<td>1.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
<td>1.0</td>
</tr>
<tr class="odd">
<td>G</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>T</td>
<td>0.0</td>
<td>0.5</td>
<td>0.5</td>
<td>0.5</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>Based on the PPM representation, all three of CAAAACC, CTTTTCC, and CTATACC are equally likely. Though looking at the starting sequences, should CTATACC really be considered so? For transcription factor binding sites, generally the answer is no. By incorporating this type of information into the motif, this can allow for increased accuracy in motif searching. A few implementations of this include: TFFM by <span class="citation">Mathelier and Wasserman (2013)</span>, BaMM by <span class="citation">Siebert and Soding (2016)</span>, and KSM by <span class="citation">Guo et al. (2018)</span>.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package implements its’ own, much simplified, version of this concept. Plainly, the standard PPM has been extended to include k-letter frequencies, with k being any number higher than 1. For example, the 2-letter version of the table @ref(tab:ppm2) motif would be:</p>
<table class="table">
<caption>(#tab:multi) 2-letter probability matrix.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AA</td>
<td>0.0</td>
<td>0.5</td>
<td>0.5</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>AC</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.5</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td>AG</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>AT</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td>CA</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>CC</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.0</td>
</tr>
<tr class="odd">
<td>CG</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>CT</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td>GA</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>GC</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td>GG</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>GT</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td>TA</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>TC</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.5</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td>TG</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>TT</td>
<td>0.0</td>
<td>0.5</td>
<td>0.5</td>
<td>0.5</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>
<p>This format shows the probability of each letter combined with the probability of the letter in the next position. The seventh column has been dropped, since it is not needed; the information in the sixth column is sufficient, and there is no eighth position to draw 2-letter probabilities from. Now, the probability of getting CTATACC is no longer equal to CTTTTCC and CAAAACC. This information is kept in the <code>multifreq</code> slot of <code>universalmotif</code> class motifs. To add this information, use the <code>add_multifreq</code> function.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">motif &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_motif.html">create_motif</a></span>(<span class="st">"CWWWWCC"</span>, <span class="dt">nsites =</span> <span class="dv">6</span>)  <span class="co">## -&gt; see next section</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">sequences &lt;-<span class="st"> </span>Biostrings<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/Biostrings/topics/XStringSet-class">DNAStringSet</a></span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">"CAAAACC"</span>, <span class="st">"CTTTTCC"</span>), <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">motif.k2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/add_multifreq.html">add_multifreq</a></span>(motif, sequences, <span class="dt">add.k =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">motif.k2</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">#&gt;          Total IC:   10</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co">#&gt;         Consensus:   CWWWWCC</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co">#&gt;      Target sites:   6</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co">#&gt;    k-letter freqs:   2 </span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="co">#&gt;   C   W   W   W   W C C</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16"><span class="co">#&gt; A 0 0.5 0.5 0.5 0.5 0 0</span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="co">#&gt; C 1 0.0 0.0 0.0 0.0 1 1</span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="co">#&gt; G 0 0.0 0.0 0.0 0.0 0 0</span></a>
<a class="sourceLine" id="cb1-19" data-line-number="19"><span class="co">#&gt; T 0 0.5 0.5 0.5 0.5 0 0</span></a></code></pre></div>
</div>
<div id="motif-trees-with-ggtree" class="section level2">
<h2 class="hasAnchor">
<a href="#motif-trees-with-ggtree" class="anchor"></a>Motif trees with ggtree</h2>
<p>Additionally, this package introduces the <code>motif_tree</code> function for generating basic tree-like diagrams for comparing motifs.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># library(MotifDb)</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co">## motif_tree is much simpler, but also much faster and allows for</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">## bigger trees; done via the ggtree package</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="kw"><a href="../reference/motif_tree.html">motif_tree</a></span>(MotifDb[<span class="dv">301</span><span class="op">:</span><span class="dv">400</span>], <span class="dt">layout =</span> <span class="st">"daylight"</span>, <span class="dt">linecol =</span> <span class="st">"family"</span>)</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">#&gt; Comparing motifs...</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98%100%</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#&gt; Average angle change [1] 0.080931235825627</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#&gt; Average angle change [2] 0.02439105682278</span></a></code></pre></div>
<p><img src="AdvancedUsage_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>#&gt; R version 3.5.1 (2018-07-02)
#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)
#&gt; Running under: macOS High Sierra 10.13.4
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
#&gt; [8] methods   base     
#&gt; 
#&gt; other attached packages:
#&gt;  [1] bindrcpp_0.2.2        MotifDb_1.23.0        Biostrings_2.49.1    
#&gt;  [4] XVector_0.21.3        IRanges_2.15.17       S4Vectors_0.19.19    
#&gt;  [7] BiocGenerics_0.27.1   universalmotif_0.99.0 BiocStyle_2.9.6      
#&gt; [10] nvimcom_0.9-75       
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] Biobase_2.41.2              tidyr_0.8.1                
#&gt;  [3] jsonlite_1.5                gtools_3.8.1               
#&gt;  [5] Rdpack_0.9-1                assertthat_0.2.0           
#&gt;  [7] BiocManager_1.30.2          rvcheck_0.1.0              
#&gt;  [9] GenomeInfoDbData_1.1.0      Rsamtools_1.33.5           
#&gt; [11] yaml_2.2.0                  pillar_1.3.0               
#&gt; [13] backports_1.1.2             lattice_0.20-35            
#&gt; [15] glue_1.3.0                  digest_0.6.17              
#&gt; [17] GenomicRanges_1.33.13       colorspace_1.3-2           
#&gt; [19] Matrix_1.2-14               htmltools_0.3.6            
#&gt; [21] plyr_1.8.4                  XML_3.98-1.16              
#&gt; [23] pkgconfig_2.0.2             bibtex_0.4.2               
#&gt; [25] ggseqlogo_0.1               bookdown_0.7               
#&gt; [27] zlibbioc_1.27.0             purrr_0.2.5                
#&gt; [29] tidytree_0.1.9              scales_1.0.0               
#&gt; [31] processx_3.2.0              BiocParallel_1.15.12       
#&gt; [33] tibble_1.4.2                ggplot2_3.0.0              
#&gt; [35] SummarizedExperiment_1.11.6 lazyeval_0.2.1             
#&gt; [37] splitstackshape_1.4.6       magrittr_1.5               
#&gt; [39] crayon_1.3.4                memoise_1.1.0              
#&gt; [41] evaluate_0.11               ps_1.1.0                   
#&gt; [43] fs_1.2.6                    nlme_3.1-137               
#&gt; [45] MASS_7.3-50                 xml2_1.2.0                 
#&gt; [47] data.table_1.11.4           tools_3.5.1                
#&gt; [49] matrixStats_0.54.0          gbRd_0.4-11                
#&gt; [51] stringr_1.3.1               ggtree_1.13.3              
#&gt; [53] munsell_0.5.0               DelayedArray_0.7.41        
#&gt; [55] compiler_3.5.1              pkgdown_1.1.0.9000         
#&gt; [57] GenomeInfoDb_1.17.1         rlang_0.2.2                
#&gt; [59] debugme_1.1.0               grid_3.5.1                 
#&gt; [61] RCurl_1.95-4.11             labeling_0.3               
#&gt; [63] bitops_1.0-6                rmarkdown_1.10             
#&gt; [65] gtable_0.2.0                roxygen2_6.1.0             
#&gt; [67] R6_2.2.2                    GenomicAlignments_1.17.3   
#&gt; [69] knitr_1.20                  dplyr_0.7.6                
#&gt; [71] rtracklayer_1.41.5          bindr_0.1.1                
#&gt; [73] commonmark_1.6              rprojroot_1.3-2            
#&gt; [75] treeio_1.5.5                desc_1.2.0                 
#&gt; [77] ape_5.1                     stringi_1.2.4              
#&gt; [79] Rcpp_0.12.18                tidyselect_0.2.4           
#&gt; [81] xfun_0.3</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-ksm">
<p>Guo, Y., K. Tian, H. Zeng, X. Guo, and D.K. Gifford. 2018. “A Novel K-Mer Set Memory (KSM) Motif Representation Improves Regulatory Variant Prediction.” <em>Genome Research</em> 28: 891–900.</p>
</div>
<div id="ref-tffm">
<p>Mathelier, A., and W.W. Wasserman. 2013. “The Next Generation of Transcription Factor Binding Site Prediction.” <em>PLoS Computational Biology</em> 9 (9): e1003214.</p>
</div>
<div id="ref-bamm">
<p>Siebert, M., and J. Soding. 2016. “Bayesian Markov Models Consistently Outperform PWMs at Predicting Motifs in Nucleotide Sequences.” <em>Nucleic Acids Research</em> 44 (13): 6055–69.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#multi-letter-information">Multi-letter information</a></li>
      <li><a href="#motif-trees-with-ggtree">Motif trees with ggtree</a></li>
      <li><a href="#session-info">Session info</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Benjamin Tremblay.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
