<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to sequence motifs • universalmotif</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to sequence motifs">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">universalmotif</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AdvancedUsage.html">Advanced motif and sequence manipulation</a>
    </li>
    <li>
      <a href="../articles/Introduction.html">Introduction to the "universalmotif" package</a>
    </li>
    <li>
      <a href="../articles/IntroductionToSequenceMotifs.html">Introduction to sequence motifs</a>
    </li>
    <li>
      <a href="../articles/MotifManipulation.html">Motif import, export, visualization, and manipulation</a>
    </li>
    <li>
      <a href="../articles/SequenceSearches.html">Sequence creation, manipulation, and scanning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bjmt/universalmotif">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction to sequence motifs</h1>
                        <h4 class="author">Benjamin Tremblay</h4>
            <address class="author_afil">
      University of Waterloo, Waterloo, Canada<br><a class="author_email" href="mailto:#"></a><a href="mailto:b2tremblay@uwaterloo.ca">b2tremblay@uwaterloo.ca</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/bjmt/universalmotif/blob/master/R/vignettes/IntroductionToSequenceMotifs.Rmd"><code>R/vignettes/IntroductionToSequenceMotifs.Rmd</code></a></small>
      <div class="hidden name"><code>IntroductionToSequenceMotifs.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Introducing sequence motif types.
    </div>
    
<div id="introduction-to-sequence-motifs" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-sequence-motifs" class="anchor"></a>Introduction to sequence motifs</h1>
<p>Motifs are a more practical representation of consensus elements in biological sequences, allowing for a more detailed description of the variability at each site <span class="citation">(see Stormo 2000)</span>. There are a number of ways of representing motifs; four types are available here. All of these can be stored using the same <code>universalmotif</code> class. Unfortunately, the naming conventions for the various motif types are not unanimous. In order to provide a simple interface to the various types, in this package they are referred to as Position Count Matrix (PCM), Position Probability Matrix (PPM), Position Weight Matrix (PWM), and Information Content Matrix (ICM).</p>
<p>These four types will be described here. Furthermore, the actual code used to go between these types can be see in <code><a href="../reference/utilities.html">?utilities</a></code>.</p>
<div id="position-count-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#position-count-matrices" class="anchor"></a>Position count matrices</h2>
<p>Also known as position frequency matrices, these are typically the most basic representation of a motif. Simply, for each position the total counts of letters is shown. For example, the following sequences:</p>
<table class="table">
<caption>(#tab:seqs) Starting sequences.</caption>
<thead><tr class="header">
<th>#</th>
<th>Sequence</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>AAGAAT</td>
</tr>
<tr class="even">
<td>2</td>
<td>ATCATA</td>
</tr>
<tr class="odd">
<td>3</td>
<td>AAGTAA</td>
</tr>
<tr class="even">
<td>4</td>
<td>AACAAA</td>
</tr>
<tr class="odd">
<td>5</td>
<td>ATTAAA</td>
</tr>
<tr class="even">
<td>6</td>
<td>AAGAAT</td>
</tr>
</tbody>
</table>
<p>… would be represented as:</p>
<table class="table">
<caption>(#tab:pcm) Position count matrix.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>6</td>
<td>4</td>
<td>0</td>
<td>5</td>
<td>5</td>
<td>4</td>
</tr>
<tr class="even">
<td>C</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td>G</td>
<td>0</td>
<td>0</td>
<td>3</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>T</td>
<td>0</td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>2</td>
</tr>
</tbody>
</table>
<p><em>(Note that all positions sum to 6, the initial number of sequences.)</em></p>
</div>
<div id="position-probability-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#position-probability-matrices" class="anchor"></a>Position probability matrices</h2>
<p>Also sometimes known as position frequency matrices and position weight matrices, these represent the probabilities for each letter at each position. Using the previous motif (table @ref(tab:pcm)) as an example, the following formula would be used to calculate the probability of each letter <em>N</em> at each individual position:</p>
<p><span class="math display">\[\begin{equation}
  Prob_N = \frac{Counts_N}{Counts_{Total}}
  (\#eq:pcm2ppm)
\end{equation}\]</span></p>
<p>This leads to the following motif representation:</p>
<table class="table">
<caption>(#tab:ppm) Position probability matrix.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>1.00</td>
<td>0.67</td>
<td>0.00</td>
<td>0.83</td>
<td>0.83</td>
<td>0.66</td>
</tr>
<tr class="even">
<td>C</td>
<td>0.00</td>
<td>0.00</td>
<td>0.33</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
</tr>
<tr class="odd">
<td>G</td>
<td>0.00</td>
<td>0.00</td>
<td>0.50</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
</tr>
<tr class="even">
<td>T</td>
<td>0.00</td>
<td>0.33</td>
<td>0.17</td>
<td>0.17</td>
<td>0.17</td>
<td>0.33</td>
</tr>
</tbody>
</table>
<p><em>(Note that all positions sum to 1.)</em></p>
<p>From this type of representation, the probability of any combination of letters can be calculated. For example, the probability for AAGAAA is about 15%. However, when starting from a small pool of sequences, many zeroes can appear in the PPM; meaning that, for example, the probability of AAAAAA is currently zero. When scanning through large numbers of biological sequences, throwing away combinations of letters such as these can be undesirable, as for example mimatches can be quite common in transcription factor binding sites; this can be fixed by adding a ‘pseudocount’. Usually a small number such as 1, it is introduced into the PCM to PPM caculation:</p>
<p><span class="math display">\[\begin{equation}
  Prob_N = \frac{Counts_N + \frac{Pseudocount}{length_N}}{Counts_{Total} + Pseudocount}
  (\#eq:ppmpseudo)
\end{equation}\]</span></p>
<p>In this equation, the pseudocount is added to the top and bottom of the fraction. However for the top fraction, which is specific to each letter, the pseudocount is divided by the total number of letters (in the case of DNA, 4). This then generates the following motif:</p>
<table class="table">
<caption>(#tab:ppmpseudo) Position probability matrix with a pseudocount of 1.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>0.892</td>
<td>0.610</td>
<td>0.036</td>
<td>0.750</td>
<td>0.750</td>
<td>0.610</td>
</tr>
<tr class="even">
<td>C</td>
<td>0.036</td>
<td>0.035</td>
<td>0.320</td>
<td>0.035</td>
<td>0.035</td>
<td>0.035</td>
</tr>
<tr class="odd">
<td>G</td>
<td>0.036</td>
<td>0.035</td>
<td>0.464</td>
<td>0.035</td>
<td>0.035</td>
<td>0.035</td>
</tr>
<tr class="even">
<td>T</td>
<td>0.036</td>
<td>0.320</td>
<td>0.180</td>
<td>0.180</td>
<td>0.180</td>
<td>0.320</td>
</tr>
</tbody>
</table>
<p>Now, though unlikely, it is no longer considered impossible for the sequence AAAAAA to exist as part of this motif. Since the total number of sequences in this case is quite low, the pseudocount can have a large impact on the values within the matrix; this impact decreases as the number of sequences increases. Of course, this can also be changed by using different pseudocounts.</p>
</div>
<div id="position-weight-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#position-weight-matrices" class="anchor"></a>Position weight matrices</h2>
<p>The position weight matrix, also known as position-specific weight matrix, position-specific scoring matrix, and logodds scoring matrix, was first proposed by <span class="citation">Stormo et al. (1982)</span>. In this case for each position, every letter has a ‘score’; this can be used to evaluate how well a sequence matches a motif. Though there can be multiple ways of calculating these scores, the most common method is to calculate the log of each probability, correcting for background frequencies. This results in the following calculation:</p>
<p><span class="math display">\[\begin{equation}
  Score_N = \log_{2}\left(\frac{Prob_N}{ProbBkg_N}\right)
  (\#eq:ppm2pwm)
\end{equation}\]</span></p>
<p>Using this equation, the log of fractions where the probability of a certain letter in a sequence is higher than that of the background probability of that letter result in positive scores, and vice versa for negative scores. Using the table @ref(tab:ppm) motif and assuming a uniform background frequency (i.e. the probability of each of the four letters is 0.25), this results in the following PWM:</p>
<table class="table">
<caption>(#tab:pwm) Position weight matrix.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>2</td>
<td>1.425</td>
<td>-Inf</td>
<td>1.737</td>
<td>1.737</td>
<td>1.415</td>
</tr>
<tr class="even">
<td>C</td>
<td>-Inf</td>
<td>-Inf</td>
<td>0.415</td>
<td>-Inf</td>
<td>-Inf</td>
<td>-Inf</td>
</tr>
<tr class="odd">
<td>G</td>
<td>-Inf</td>
<td>-Inf</td>
<td>1.000</td>
<td>-Inf</td>
<td>-Inf</td>
<td>-Inf</td>
</tr>
<tr class="even">
<td>T</td>
<td>-Inf</td>
<td>0.415</td>
<td>-0.585</td>
<td>-0.585</td>
<td>-0.595</td>
<td>0.415</td>
</tr>
</tbody>
</table>
<p><em>(Note that the position totals no longer have equal sums.)</em></p>
<p>In order to score a sequence, add up the score for the letters at the specific positions; for example AAGAAA has a score of 9.31. However, similar to with PPMs, if starting from a small pool of sequences the sequence AAAAAA could never be recovered using this motif model, with a score of -Inf. This can be avoided simply by starting from a pseudocount-adjusted PPM. Using the table @ref(tab:ppmpseudo) motif, this becomes:</p>
<table class="table">
<caption>(#tab:pwmpseudo) Position weight matrix with a pseudocount of 1.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>1.840</td>
<td>1.280</td>
<td>-2.807</td>
<td>1.585</td>
<td>1.585</td>
<td>1.280</td>
</tr>
<tr class="even">
<td>C</td>
<td>-2.807</td>
<td>-2.807</td>
<td>0.363</td>
<td>-2.807</td>
<td>-2.807</td>
<td>-2.807</td>
</tr>
<tr class="odd">
<td>G</td>
<td>-2.807</td>
<td>-2.807</td>
<td>0.893</td>
<td>-2.807</td>
<td>-2.807</td>
<td>-2.807</td>
</tr>
<tr class="even">
<td>T</td>
<td>-2.807</td>
<td>0.363</td>
<td>-0.485</td>
<td>-0.485</td>
<td>-0.485</td>
<td>0.363</td>
</tr>
</tbody>
</table>
<p>Now, the score for AAGAA is 8.46, and the score for AAAAAA is 4.76. Though the score for the latter is low, it is no longer -Inf as a result of one mismatch.</p>
<p>When searching for instances of this motif using this scoring system, one more consideration is needed: a minimum score. For example, both sequences AAGAAA and ATTTTT have positive scores; but one is much higher than the other. Should ATTTTT then be discarded? In order to answer this a threshold is set; typically this is a certain percent of the highest possible score. For example, in the table @ref(tab:pwmpseudo) motif, the highest possible score is 8.46; using a threshold of 25%, the minimum score then becomes 2.115. This means that ATTTTT, with a score of 1.112, would indeed be discarded.</p>
<p>Other methods for determining the minimum score include starting from P-values. The package <em><a href="https://CRAN.R-project.org/package=TFMPvalue">TFMPvalue</a></em> for example can calculate minimum scores from P-values, using the algorithm described by <span class="citation">Touzet and Varre (2007)</span>.</p>
</div>
<div id="information-content-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#information-content-matrices" class="anchor"></a>Information content matrices</h2>
<p>Finally, the information content matrix <span class="citation">(Schneider et al. 1986; Schneider and Stephens 1990)</span>. This type aims to include another consideration: are some positions more important than others? To explore this, let us consider the table @ref(tab:ppm) motif. This matrix can be represented as a sequence logo using the <em><a href="https://bioconductor.org/packages/3.8/seqLogo">seqLogo</a></em> package:</p>
<pre><code>#&gt; Loading required package: grid</code></pre>
<div class="figure">
<img src="IntroductionToSequenceMotifs_files/figure-html/unnamed-chunk-1-1.png" alt="Sequence logo of a PPM" width="700"><p class="caption">
Sequence logo of a PPM
</p>
</div>
<p>In this case, the height of the letters represent their probabilities at each position. However, when represented as an ICM, the sequence logo then resembles:</p>
<div class="figure">
<img src="IntroductionToSequenceMotifs_files/figure-html/unnamed-chunk-2-1.png" alt="Sequence logo of an ICM" width="700"><p class="caption">
Sequence logo of an ICM
</p>
</div>
<p>Now, the total height of each position is scaled using the total <em>information</em> at that position. Simply put, the total information of each position is an indication of the level of conservation; in the example motif, the first position is highly conserved, always being the letter A, whereas the third position is less so conserved, as the probabilities for any one letter are quite lower.</p>
<p>For every position, the letter heights are calculated as the total possible information content (IC) or maximum uncertainty, then subtracting that positions’ actual uncertainty <span class="citation">(Schneider et al. 1986; Schneider and Stephens 1990)</span>. This calculation is based on Shannon’s entropy <span class="citation">(Shannon 1948)</span>, with the final values representing ‘bits’ <span class="citation">(Schneider 1991)</span>. The total IC is based on alphabet length, using the following equation:</p>
<p><span class="math display">\[\begin{equation}
  IC_{total} = \log_{2}(length_N)
  (\#eq:totalic)
\end{equation}\]</span></p>
<p>For DNA motifs which have an alphabet length of 4, the total IC is 2. To calculate the positions’ actual uncertainty, the following equation is used:</p>
<p><span class="math display">\[\begin{equation}
  Uncertainty = - \sum_{N=A}^{T} Prob_N \times{} \log_{2}(Prob_N)
  (\#eq:posuncer)
\end{equation}\]</span></p>
<p>Where the position uncertainty is the sum of the uncertainty of all alphabet letters (A, C, G, and T). To calculate the final information content:</p>
<p><span class="math display">\[\begin{equation}
  IC_{position} = IC_{total} - Uncertainty
  (\#eq:ic)
\end{equation}\]</span></p>
<p>In the original implementation described by <span class="citation">Schneider et al. (1986)</span>, an additional error correction factor is included to account for sample size. This correction is rarely used however, but can be incorporated as such:</p>
<p><span class="math display">\[\begin{equation}
  IC_{position} = IC_{total} - [ Uncertainty + E_{corr} ]
  (\#eq:iccorr)
\end{equation}\]</span></p>
<p>The details for calculating this factor will not be covered here; refer to <span class="citation">Schneider et al. (1986)</span>. The <em><a href="https://bioconductor.org/packages/3.8/TFBSTools">TFBSTools</a></em> package offers the ability to incorporate this error correction.</p>
<p>Finally, to get the height of each letter at each position, the final IC is multiplied by the letter and position probabilities:</p>
<p><span class="math display">\[\begin{equation}
  IC_N = Prob_N \times{} IC_{position}
  (\#eq:finalic)
\end{equation}\]</span></p>
<p>Using the above equations, the table @ref(tab:ppm) motif then becomes:</p>
<table class="table">
<caption>(#tab:icm) Information content matrix.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>2.000</td>
<td>0.721</td>
<td>0.000</td>
<td>1.125</td>
<td>1.125</td>
<td>0.721</td>
</tr>
<tr class="even">
<td>C</td>
<td>0.000</td>
<td>0.000</td>
<td>0.180</td>
<td>0.000</td>
<td>0.000</td>
<td>0.000</td>
</tr>
<tr class="odd">
<td>G</td>
<td>0.000</td>
<td>0.000</td>
<td>0.270</td>
<td>0.000</td>
<td>0.000</td>
<td>0.000</td>
</tr>
<tr class="even">
<td>T</td>
<td>0.000</td>
<td>0.361</td>
<td>0.090</td>
<td>0.225</td>
<td>0.225</td>
<td>0.361</td>
</tr>
</tbody>
</table>
<p><em>(Note that none of the positions have a sum larger than 2.)</em></p>
<p>An alternative to representing information content as Shannon’s entropy is relative entropy, or Kullback-Leibler divergence <span class="citation">(Kullback and Leibler 1951)</span>. While IC as Shannon’s entropy has the advantage of having a consistent maximum IC for every position, it does not take into account non-uniform background frequencies. Relative entropy on the other hand will take this into account, but all positions no longer share the same maximum IC. To calculate relative entropy:</p>
<p><span class="math display">\[\begin{equation}
  IC_N = Prob_N \times{} \log_{2}\left(\frac{Prob_N}{ProbBkg_N}\right)
  (\#eq:kldiv)
\end{equation}\]</span></p>
<p>Using this equation can lead to IC less than zero; these values are not allowed, so they are simply replaced with zero. With this equation and assuming uniform background frequencies, the table @ref(tab:ppm) motif becomes:</p>
<table class="table">
<caption>(#tab:icmkl) Information content matrix as relative divergence.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>1.640</td>
<td>0.777</td>
<td>0.000</td>
<td>1.190</td>
<td>1.190</td>
<td>0.777</td>
</tr>
<tr class="even">
<td>C</td>
<td>0.000</td>
<td>0.000</td>
<td>0.177</td>
<td>0.000</td>
<td>0.000</td>
<td>0.000</td>
</tr>
<tr class="odd">
<td>G</td>
<td>0.000</td>
<td>0.000</td>
<td>0.415</td>
<td>0.000</td>
<td>0.000</td>
<td>0.000</td>
</tr>
<tr class="even">
<td>T</td>
<td>0.000</td>
<td>0.177</td>
<td>0.090</td>
<td>0.000</td>
<td>0.000</td>
<td>0.117</td>
</tr>
</tbody>
</table>
<p>This motif would look significantly different with non-uniform background frequences. For example, starting from the following background frequencies: <code>c(A = 0.4, C = 0.1, G = 0.1, T= 0.4)</code>, the motif resembles:</p>
<table class="table">
<caption>(#tab:icmkl2) Information content matrix as relative divergence with a non-uniform background.</caption>
<thead><tr class="header">
<th>Position</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td>1.030</td>
<td>0.366</td>
<td>0.000</td>
<td>0.680</td>
<td>0.680</td>
<td>0.366</td>
</tr>
<tr class="even">
<td>C</td>
<td>0.000</td>
<td>0.000</td>
<td>0.541</td>
<td>0.000</td>
<td>0.000</td>
<td>0.000</td>
</tr>
<tr class="odd">
<td>G</td>
<td>0.000</td>
<td>0.000</td>
<td>1.028</td>
<td>0.000</td>
<td>0.000</td>
<td>0.000</td>
</tr>
<tr class="even">
<td>T</td>
<td>0.000</td>
<td>0.000</td>
<td>0.090</td>
<td>0.000</td>
<td>0.000</td>
<td>0.000</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>#&gt; R version 3.5.1 (2018-07-02)
#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)
#&gt; Running under: macOS High Sierra 10.13.4
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] grid      stats     graphics  grDevices utils     datasets  methods  
#&gt; [8] base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] seqLogo_1.47.0        universalmotif_0.99.0 BiocStyle_2.9.6      
#&gt; [4] nvimcom_0.9-75       
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] Rcpp_0.12.18        ape_5.1             lattice_0.20-35    
#&gt;  [4] tidyr_0.8.1         ps_1.1.0            Biostrings_2.49.1  
#&gt;  [7] ggseqlogo_0.1       gtools_3.8.1        assertthat_0.2.0   
#&gt; [10] rprojroot_1.3-2     digest_0.6.17       R6_2.2.2           
#&gt; [13] plyr_1.8.4          backports_1.1.2     stats4_3.5.1       
#&gt; [16] evaluate_0.11       highr_0.7           ggplot2_3.0.0      
#&gt; [19] pillar_1.3.0        Rdpack_0.9-1        zlibbioc_1.27.0    
#&gt; [22] rlang_0.2.2         lazyeval_0.2.1      S4Vectors_0.19.19  
#&gt; [25] rmarkdown_1.10      pkgdown_1.1.0.9000  desc_1.2.0         
#&gt; [28] stringr_1.3.1       munsell_0.5.0       compiler_3.5.1     
#&gt; [31] xfun_0.3            pkgconfig_2.0.2     BiocGenerics_0.27.1
#&gt; [34] htmltools_0.3.6     tidyselect_0.2.4    tibble_1.4.2       
#&gt; [37] roxygen2_6.1.0      bookdown_0.7        IRanges_2.15.17    
#&gt; [40] crayon_1.3.4        dplyr_0.7.6         MASS_7.3-50        
#&gt; [43] commonmark_1.6      nlme_3.1-137        jsonlite_1.5       
#&gt; [46] gtable_0.2.0        magrittr_1.5        scales_1.0.0       
#&gt; [49] bibtex_0.4.2        tidytree_0.1.9      stringi_1.2.4      
#&gt; [52] debugme_1.1.0       XVector_0.21.3      fs_1.2.6           
#&gt; [55] bindrcpp_0.2.2      ggtree_1.13.3       xml2_1.2.0         
#&gt; [58] rvcheck_0.1.0       tools_3.5.1         treeio_1.5.5       
#&gt; [61] glue_1.3.0          purrr_0.2.5         processx_3.2.0     
#&gt; [64] parallel_3.5.1      yaml_2.2.0          colorspace_1.3-2   
#&gt; [67] BiocManager_1.30.2  gbRd_0.4-11         memoise_1.1.0      
#&gt; [70] knitr_1.20          bindr_0.1.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-kl">
<p>Kullback, S., and R. A. Leibler. 1951. “On Information and Sufficiency.” <em>The Annals of Mathematical Statistics</em> 22: 79–86.</p>
</div>
<div id="ref-bits">
<p>Schneider, T. D. 1991. “Theory of Molecular Machines. II. Energy Dissipation from Molecular Machines.” <em>Journal of Theoretical Biology</em> 148 (1): 125–37.</p>
</div>
<div id="ref-icm">
<p>Schneider, T. D., and R. M. Stephens. 1990. “Sequence Logos: A New Way to Display Consensus Sequences.” <em>Nucleic Acids Research</em> 18 (20): 6097–6100.</p>
</div>
<div id="ref-correction">
<p>Schneider, Thomas D., Gary D. Stormo, Larry Gold, and Andrzej Ehrenfeucht. 1986. “Information Content of Binding Sites on Nucleotide Sequences.” <em>Journal of Molecular Biology</em> 188 (3): 415–31.</p>
</div>
<div id="ref-shannon">
<p>Shannon, Claude E. 1948. “A Mathematical Theory of Communication.” <em>Bell System Technical Journal</em> 27 (3): 379–423.</p>
</div>
<div id="ref-pwm">
<p>Stormo, Gary D., Thomas D. Schneider, Larry Gold, and Andrzej Ehrenfeucht. 1982. “Use of the Perceptron Algorithm to Distinguish Translational Initiation Sites in E. coli.” <em>Nucleic Acids Research</em> 10 (9): 2997–3011.</p>
</div>
<div id="ref-motifhist">
<p>Stormo, G. D. 2000. “DNA Binding Sites: Representation and Discovery.” <em>Bioinformatics</em> 16 (1): 16–23.</p>
</div>
<div id="ref-pval">
<p>Touzet, H., and J.-S. Varre. 2007. “Efficient and Accurate P-Value Computation for Position Weight Matrices.” <em>Algorithms for Molecular Biology</em> 2: 15.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction-to-sequence-motifs">Introduction to sequence motifs</a><ul class="nav nav-pills nav-stacked">
<li><a href="#position-count-matrices">Position count matrices</a></li>
      <li><a href="#position-probability-matrices">Position probability matrices</a></li>
      <li><a href="#position-weight-matrices">Position weight matrices</a></li>
      <li><a href="#information-content-matrices">Information content matrices</a></li>
      </ul>
</li>
      <li><a href="#session-info">Session info</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Benjamin Tremblay.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
