<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Motif import, export, visualization, and manipulation • universalmotif</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Motif import, export, visualization, and manipulation" />

<meta property="og:description" content="" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-article">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">universalmotif</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/AdvancedUsage.html">Advanced motif and sequence manipulation</a>
    </li>
    <li>
      <a href="../articles/Introduction.html">Introduction to the "universalmotif" package</a>
    </li>
    <li>
      <a href="../articles/IntroductionToSequenceMotifs.html">Introduction to sequence motifs</a>
    </li>
    <li>
      <a href="../articles/MotifManipulation.html">Motif import, export, visualization, and manipulation</a>
    </li>
    <li>
      <a href="../articles/SequenceSearches.html">Sequence creation, manipulation, and scanning</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bjmt/universalmotif">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>


<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Motif import, export, visualization, and manipulation</h1>
                        <h4 class="author">Benjamin Tremblay</h4>
            <address class="author_afil">
      University of Waterloo, Waterloo, Canada<br>            <a class="author_email" href="mailto:#"><a href="mailto:b2tremblay@uwaterloo.ca">b2tremblay@uwaterloo.ca</a></a>
      </address>
                  
      
      <small class="dont-index">Source: <a href='https://github.com/bjmt/universalmotif/blob/master/R/vignettes/MotifManipulation.Rmd'><code>R/vignettes/MotifManipulation.Rmd</code></a></small>
      <div class="hidden name"><code>MotifManipulation.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Introducing sequence motif types.
    </div>
    
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
##     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
##     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which, which.max,
##     which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## 
## Attaching package: &#39;Biostrings&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     strsplit</code></pre>
<pre><code>## See system.file(&quot;LICENSE&quot;, package=&quot;MotifDb&quot;) for use restrictions.</code></pre>
<div id="the-universalmotif-class-and-conversion-utilities" class="section level1">
<h1>The universalmotif class and conversion utilities</h1>
<div id="the-universalmotif-class" class="section level2">
<h2>The universalmotif class</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package stores motifs using the <code>universalmotif</code> class. The most basic <code>universalmotif</code> object exposes the ‘name’, ‘alphabet’, ‘type’, ‘type’, ‘strand’, ‘icscore’, ‘consensus’, and ‘motif’ slots; furthermore, the ‘pseudocount’ and ‘bkg’ slots are also stored but not shown. <code>universalmotif</code> class motifs can be PCM, PPM, PWM, or ICM type.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw">data</span>(examplemotif)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3">examplemotif</a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb17-6" data-line-number="6"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb17-7" data-line-number="7"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb17-8" data-line-number="8"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb17-9" data-line-number="9"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb17-10" data-line-number="10"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb17-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb17-12" data-line-number="12"><span class="co">#&gt;   T A T A   W A   W</span></a>
<a class="sourceLine" id="cb17-13" data-line-number="13"><span class="co">#&gt; A 0 1 0 1 0.5 1 0.5</span></a>
<a class="sourceLine" id="cb17-14" data-line-number="14"><span class="co">#&gt; C 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb17-15" data-line-number="15"><span class="co">#&gt; G 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb17-16" data-line-number="16"><span class="co">#&gt; T 1 0 1 0 0.5 0 0.5</span></a></code></pre></div>
<p>Exposing the motif class with <code>str</code> shows the other available slots:</p>
<pre><code>#&gt; Formal class &#39;universalmotif&#39; [package &quot;universalmotif&quot;] with 19 slots
#&gt;   ..@ name       : chr &quot;motif&quot;
#&gt;   ..@ altname    : chr(0) 
#&gt;   ..@ family     : chr(0) 
#&gt;   ..@ organism   : chr(0) 
#&gt;   ..@ motif      : num [1:4, 1:7] 0 0 0 1 1 0 0 0 0 0 ...
#&gt;   .. ..- attr(*, &quot;dimnames&quot;)=List of 2
#&gt;   .. .. ..$ : chr [1:4] &quot;A&quot; &quot;C&quot; &quot;G&quot; &quot;T&quot;
#&gt;   .. .. ..$ : chr [1:7] &quot;T&quot; &quot;A&quot; &quot;T&quot; &quot;A&quot; ...
#&gt;   ..@ alphabet   : chr &quot;DNA&quot;
#&gt;   ..@ type       : chr &quot;PPM&quot;
#&gt;   ..@ icscore    : num 14
#&gt;   ..@ nsites     : num(0) 
#&gt;   ..@ pseudocount: num 0
#&gt;   ..@ bkg        : num [1:4] 0.25 0.25 0.25 0.25
#&gt;   ..@ bkgsites   : num(0) 
#&gt;   ..@ consensus  : chr &quot;TATAWAW&quot;
#&gt;   ..@ strand     : chr &quot;+-&quot;
#&gt;   ..@ pval       : num(0) 
#&gt;   ..@ qval       : num(0) 
#&gt;   ..@ eval       : num(0) 
#&gt;   ..@ multifreq  : list()
#&gt;   ..@ extrainfo  : chr(0)</code></pre>
<p>The other slots will be shown once filled:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw">data</span>(examplemotif)</a>
<a class="sourceLine" id="cb19-3" data-line-number="3"></a>
<a class="sourceLine" id="cb19-4" data-line-number="4"><span class="co">## The various slots can be accessed individually using `[`</span></a>
<a class="sourceLine" id="cb19-5" data-line-number="5">examplemotif[<span class="st">&quot;consensus&quot;</span>]</a>
<a class="sourceLine" id="cb19-6" data-line-number="6"><span class="co">#&gt; consensus </span></a>
<a class="sourceLine" id="cb19-7" data-line-number="7"><span class="co">#&gt; &quot;TATAWAW&quot;</span></a>
<a class="sourceLine" id="cb19-8" data-line-number="8"></a>
<a class="sourceLine" id="cb19-9" data-line-number="9"><span class="co">## To change a slot, use `[&lt;-`</span></a>
<a class="sourceLine" id="cb19-10" data-line-number="10">examplemotif[<span class="st">&quot;family&quot;</span>] &lt;-<span class="st"> &quot;My motif family&quot;</span></a>
<a class="sourceLine" id="cb19-11" data-line-number="11">examplemotif</a>
<a class="sourceLine" id="cb19-12" data-line-number="12"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb19-13" data-line-number="13"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb19-14" data-line-number="14"><span class="co">#&gt;            Family:   My motif family</span></a>
<a class="sourceLine" id="cb19-15" data-line-number="15"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb19-16" data-line-number="16"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb19-17" data-line-number="17"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb19-18" data-line-number="18"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb19-19" data-line-number="19"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb19-20" data-line-number="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb19-21" data-line-number="21"><span class="co">#&gt;   T A T A   W A   W</span></a>
<a class="sourceLine" id="cb19-22" data-line-number="22"><span class="co">#&gt; A 0 1 0 1 0.5 1 0.5</span></a>
<a class="sourceLine" id="cb19-23" data-line-number="23"><span class="co">#&gt; C 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb19-24" data-line-number="24"><span class="co">#&gt; G 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb19-25" data-line-number="25"><span class="co">#&gt; T 1 0 1 0 0.5 0 0.5</span></a></code></pre></div>
<p>Though the slots can easily be changed manually with <code>[&lt;-</code>, a number of safeguards have been put in place for some of the slots which will prevent incorrect values from being introduced.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw">data</span>(examplemotif)</a>
<a class="sourceLine" id="cb20-3" data-line-number="3"></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"><span class="co">## The consensus slot is dependent on the motif matrix</span></a>
<a class="sourceLine" id="cb20-5" data-line-number="5">examplemotif[<span class="st">&quot;consensus&quot;</span>]</a>
<a class="sourceLine" id="cb20-6" data-line-number="6"><span class="co">#&gt; consensus </span></a>
<a class="sourceLine" id="cb20-7" data-line-number="7"><span class="co">#&gt; &quot;TATAWAW&quot;</span></a>
<a class="sourceLine" id="cb20-8" data-line-number="8"></a>
<a class="sourceLine" id="cb20-9" data-line-number="9"><span class="co">## Changing this would mean it no longer matches the motif</span></a>
<a class="sourceLine" id="cb20-10" data-line-number="10">examplemotif[<span class="st">&quot;consensus&quot;</span>] &lt;-<span class="st"> &quot;GGGAGAG&quot;</span></a>
<a class="sourceLine" id="cb20-11" data-line-number="11"><span class="co">#&gt; Error in .local(x, i, ..., value): consensus string for DNA motifs is generated automatically</span></a>
<a class="sourceLine" id="cb20-12" data-line-number="12"></a>
<a class="sourceLine" id="cb20-13" data-line-number="13"><span class="co">## Another example:</span></a>
<a class="sourceLine" id="cb20-14" data-line-number="14">examplemotif[<span class="st">&quot;strand&quot;</span>] &lt;-<span class="st"> &quot;x&quot;</span></a>
<a class="sourceLine" id="cb20-15" data-line-number="15"><span class="co">#&gt; Error in validObject(x): invalid class &quot;universalmotif&quot; object: motif &#39;strand&#39; must be either &#39;+&#39;, &#39;-&#39;, or &#39;+-&#39;</span></a>
<a class="sourceLine" id="cb20-16" data-line-number="16"></a>
<a class="sourceLine" id="cb20-17" data-line-number="17"><span class="co">## Of course, this can be circumvented by simply using `@`; do this at your</span></a>
<a class="sourceLine" id="cb20-18" data-line-number="18"><span class="co">## own risk. Trying to use a motif with an incorrectly filled slot will likely</span></a>
<a class="sourceLine" id="cb20-19" data-line-number="19"><span class="co">## cause other universalmotif functions to fail.</span></a>
<a class="sourceLine" id="cb20-20" data-line-number="20">examplemotif<span class="op">@</span>strand &lt;-<span class="st"> &quot;x&quot;</span></a>
<a class="sourceLine" id="cb20-21" data-line-number="21">examplemotif</a>
<a class="sourceLine" id="cb20-22" data-line-number="22"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb20-23" data-line-number="23"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb20-24" data-line-number="24"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb20-25" data-line-number="25"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb20-26" data-line-number="26"><span class="co">#&gt;           Strands:   x</span></a>
<a class="sourceLine" id="cb20-27" data-line-number="27"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb20-28" data-line-number="28"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb20-29" data-line-number="29"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb20-30" data-line-number="30"><span class="co">#&gt;   T A T A   W A   W</span></a>
<a class="sourceLine" id="cb20-31" data-line-number="31"><span class="co">#&gt; A 0 1 0 1 0.5 1 0.5</span></a>
<a class="sourceLine" id="cb20-32" data-line-number="32"><span class="co">#&gt; C 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb20-33" data-line-number="33"><span class="co">#&gt; G 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb20-34" data-line-number="34"><span class="co">#&gt; T 1 0 1 0 0.5 0 0.5</span></a></code></pre></div>
<p>Below the exposed metadata slots, the actual ‘motif’ matrix is shown. Each position is its’ own column; row names showing the alphabet letters, and the column names showing the consensus letter at each position.</p>
</div>
<div id="converting-to-and-from-another-packages-class" class="section level2">
<h2>Converting to and from another package’s class</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package aims to unify most of the motif-related Bioconductor packages by providing the <code>convert_motif</code> function. This allows for easy transition between supported packages (see <code>?convert_motif</code> for a complete list of supported packages).</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">data</span>(examplemotif)</a>
<a class="sourceLine" id="cb21-3" data-line-number="3"><span class="co"># data(MA0003.2)</span></a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="co"># suppressPackageStartupMessages(library(MotifDb))</span></a>
<a class="sourceLine" id="cb21-5" data-line-number="5"></a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="co">## convert from a universalmotif to another</span></a>
<a class="sourceLine" id="cb21-7" data-line-number="7"><span class="co"># convert_motifs(examplemotif, &quot;TFBSTools-PWMatrix&quot;)</span></a>
<a class="sourceLine" id="cb21-8" data-line-number="8"></a>
<a class="sourceLine" id="cb21-9" data-line-number="9"><span class="co">## convert to universalmotif</span></a>
<a class="sourceLine" id="cb21-10" data-line-number="10"><span class="co"># convert_motifs(MA0003.2)</span></a>
<a class="sourceLine" id="cb21-11" data-line-number="11"></a>
<a class="sourceLine" id="cb21-12" data-line-number="12"><span class="co">## convert between two packages</span></a>
<a class="sourceLine" id="cb21-13" data-line-number="13"><span class="co"># convert_motifs(MotifDb[1], &quot;TFBSTools-ICMatrix&quot;)</span></a></code></pre></div>
<p>The <code>convert_motifs</code> function is embedded in most of the <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> functions, meaning that compatible motif classes from other packages can be used without needed to convert them first.</p>
</div>
</div>
<div id="importing-and-exporting-motifs" class="section level1">
<h1>Importing and exporting motifs</h1>
<div id="importing" class="section level2">
<h2>Importing</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package offers a number of <code>read_</code> functions to allow for easy import of various motif formats. These include:</p>
<ul>
<li><code>read_cisbp</code>: CIS-BP <span class="citation">(Weirauch et al. 2014)</span></li>
<li><code>read_homer</code>: HOMER <span class="citation">(Heinz et al. 2010)</span></li>
<li><code>read_jaspar</code>: JASPAR <span class="citation">(Khan et al. 2018)</span></li>
<li><code>read_meme</code>: MEME <span class="citation">(Bailey et al. 2009)</span></li>
<li><code>read_transfac</code>: TRANSFAC <span class="citation">(Wingender et al. 1996)</span></li>
<li><code>read_uniprobe</code>: UniPROBE <span class="citation">(Hume et al. 2015)</span></li>
</ul>
<p>These functions should work natively with these formats, but if you are generating your own motifs in one of these formats than it must adhere quite strictly to the format. An example of each of these is included in this package; see <code>system.file(&quot;extdata&quot;, package=&quot;universalmotif&quot;)</code>.</p>
<p>Additionally, the generic <code>read_matrix</code> function can read simply-formatted motif files; each motif can have a header and a motif matrix.</p>
</div>
<div id="exporting" class="section level2">
<h2>Exporting</h2>
<p>Compatible motif classes can be written to disk using:</p>
<ul>
<li><code>write_homer</code></li>
<li><code>write_jaspar</code></li>
<li><code>write_matrix</code></li>
<li><code>write_meme</code></li>
<li><code>write_transfac</code></li>
</ul>
<p>The <code>write_matrix</code> function, similar to its’ <code>read_matrix</code> counterpart, can write motifs as simple matrices with an optional header.</p>
</div>
</div>
<div id="modifying-motifs-and-related-functions" class="section level1">
<h1>Modifying motifs and related functions</h1>
<div id="converting-motif-type" class="section level2">
<h2>Converting motif type</h2>
<p>Any <code>universalmotif</code> object can transition between PCM, PPM, PWM, and ICM types seamlessly using the <code>convert_type</code> function. The only exception to this is if the ICM calculation is performed with sample correction, or as relative entropy. If this occurs, then back conversion to another type will be inaccurate.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="kw">data</span>(examplemotif)</a>
<a class="sourceLine" id="cb22-3" data-line-number="3"></a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="co">## This motif is currently a PPM:</span></a>
<a class="sourceLine" id="cb22-5" data-line-number="5">examplemotif[<span class="st">&quot;type&quot;</span>]</a>
<a class="sourceLine" id="cb22-6" data-line-number="6"><span class="co">#&gt;  type </span></a>
<a class="sourceLine" id="cb22-7" data-line-number="7"><span class="co">#&gt; &quot;PPM&quot;</span></a>
<a class="sourceLine" id="cb22-8" data-line-number="8"></a>
<a class="sourceLine" id="cb22-9" data-line-number="9"><span class="co">## When converting to PCM, the &#39;nsites&#39; slot is needed to tell it how</span></a>
<a class="sourceLine" id="cb22-10" data-line-number="10"><span class="co">## many sequences it originated from. If empty, 100 is used.</span></a>
<a class="sourceLine" id="cb22-11" data-line-number="11"><span class="kw">convert_type</span>(examplemotif, <span class="st">&quot;PCM&quot;</span>)</a>
<a class="sourceLine" id="cb22-12" data-line-number="12"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-13" data-line-number="13"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb22-14" data-line-number="14"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb22-15" data-line-number="15"><span class="co">#&gt;              Type:   PCM</span></a>
<a class="sourceLine" id="cb22-16" data-line-number="16"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb22-17" data-line-number="17"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb22-18" data-line-number="18"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb22-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-20" data-line-number="20"><span class="co">#&gt;     T   A   T   A  W   A  W</span></a>
<a class="sourceLine" id="cb22-21" data-line-number="21"><span class="co">#&gt; A   0 100   0 100 50 100 50</span></a>
<a class="sourceLine" id="cb22-22" data-line-number="22"><span class="co">#&gt; C   0   0   0   0  0   0  0</span></a>
<a class="sourceLine" id="cb22-23" data-line-number="23"><span class="co">#&gt; G   0   0   0   0  0   0  0</span></a>
<a class="sourceLine" id="cb22-24" data-line-number="24"><span class="co">#&gt; T 100   0 100   0 50   0 50</span></a>
<a class="sourceLine" id="cb22-25" data-line-number="25"></a>
<a class="sourceLine" id="cb22-26" data-line-number="26"><span class="co">## For converting to PWM, the &#39;pseudocount&#39; slot is used to determine if any</span></a>
<a class="sourceLine" id="cb22-27" data-line-number="27"><span class="co">## correction should be applied:</span></a>
<a class="sourceLine" id="cb22-28" data-line-number="28">examplemotif[<span class="st">&quot;pseudocount&quot;</span>]</a>
<a class="sourceLine" id="cb22-29" data-line-number="29"><span class="co">#&gt; pseudocount </span></a>
<a class="sourceLine" id="cb22-30" data-line-number="30"><span class="co">#&gt;           0</span></a>
<a class="sourceLine" id="cb22-31" data-line-number="31"><span class="kw">convert_type</span>(examplemotif, <span class="st">&quot;PWM&quot;</span>)</a>
<a class="sourceLine" id="cb22-32" data-line-number="32"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-33" data-line-number="33"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb22-34" data-line-number="34"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb22-35" data-line-number="35"><span class="co">#&gt;              Type:   PWM</span></a>
<a class="sourceLine" id="cb22-36" data-line-number="36"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb22-37" data-line-number="37"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb22-38" data-line-number="38"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb22-39" data-line-number="39"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-40" data-line-number="40"><span class="co">#&gt;      T    A    T    A    W    A    W</span></a>
<a class="sourceLine" id="cb22-41" data-line-number="41"><span class="co">#&gt; A -Inf    2 -Inf    2    1    2    1</span></a>
<a class="sourceLine" id="cb22-42" data-line-number="42"><span class="co">#&gt; C -Inf -Inf -Inf -Inf -Inf -Inf -Inf</span></a>
<a class="sourceLine" id="cb22-43" data-line-number="43"><span class="co">#&gt; G -Inf -Inf -Inf -Inf -Inf -Inf -Inf</span></a>
<a class="sourceLine" id="cb22-44" data-line-number="44"><span class="co">#&gt; T    2 -Inf    2 -Inf    1 -Inf    1</span></a>
<a class="sourceLine" id="cb22-45" data-line-number="45"></a>
<a class="sourceLine" id="cb22-46" data-line-number="46"><span class="co">## You can either change the &#39;pseudocount&#39; slot manually beforehand, or pass</span></a>
<a class="sourceLine" id="cb22-47" data-line-number="47"><span class="co">## one to &#39;convert_type&#39;</span></a>
<a class="sourceLine" id="cb22-48" data-line-number="48"><span class="kw">convert_type</span>(examplemotif, <span class="st">&quot;PWM&quot;</span>, <span class="dt">pseudocount =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb22-49" data-line-number="49"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-50" data-line-number="50"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb22-51" data-line-number="51"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb22-52" data-line-number="52"><span class="co">#&gt;              Type:   PWM</span></a>
<a class="sourceLine" id="cb22-53" data-line-number="53"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb22-54" data-line-number="54"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb22-55" data-line-number="55"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb22-56" data-line-number="56"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-57" data-line-number="57"><span class="co">#&gt;       T     A     T     A     W     A     W</span></a>
<a class="sourceLine" id="cb22-58" data-line-number="58"><span class="co">#&gt; A -6.66  1.99 -6.66  1.99  0.99  1.99  0.99</span></a>
<a class="sourceLine" id="cb22-59" data-line-number="59"><span class="co">#&gt; C -6.66 -6.66 -6.66 -6.66 -6.66 -6.66 -6.66</span></a>
<a class="sourceLine" id="cb22-60" data-line-number="60"><span class="co">#&gt; G -6.66 -6.66 -6.66 -6.66 -6.66 -6.66 -6.66</span></a>
<a class="sourceLine" id="cb22-61" data-line-number="61"><span class="co">#&gt; T  1.99 -6.66  1.99 -6.66  0.99 -6.66  0.99</span></a>
<a class="sourceLine" id="cb22-62" data-line-number="62"></a>
<a class="sourceLine" id="cb22-63" data-line-number="63"><span class="co">## There are a couple of additional options for ICM conversion:</span></a>
<a class="sourceLine" id="cb22-64" data-line-number="64"><span class="co">## &#39;nsize_correction&#39; and &#39;relative_entropy&#39;. The former uses the</span></a>
<a class="sourceLine" id="cb22-65" data-line-number="65"><span class="co">## TFBSTools:::schneider_correction function for sample size correction.</span></a>
<a class="sourceLine" id="cb22-66" data-line-number="66"><span class="co">## The latter uses the &#39;bkg&#39; slot to calculate information content.</span></a>
<a class="sourceLine" id="cb22-67" data-line-number="67">examplemotif[<span class="st">&quot;nsites&quot;</span>] =<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb22-68" data-line-number="68"><span class="kw">convert_type</span>(examplemotif, <span class="st">&quot;ICM&quot;</span>, <span class="dt">nsize_correction =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb22-69" data-line-number="69"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-70" data-line-number="70"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb22-71" data-line-number="71"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb22-72" data-line-number="72"><span class="co">#&gt;              Type:   ICM</span></a>
<a class="sourceLine" id="cb22-73" data-line-number="73"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb22-74" data-line-number="74"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb22-75" data-line-number="75"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb22-76" data-line-number="76"><span class="co">#&gt;      Target sites:   10</span></a>
<a class="sourceLine" id="cb22-77" data-line-number="77"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-78" data-line-number="78"><span class="co">#&gt;   T A T A   W A   W</span></a>
<a class="sourceLine" id="cb22-79" data-line-number="79"><span class="co">#&gt; A 0 2 0 2 0.5 2 0.5</span></a>
<a class="sourceLine" id="cb22-80" data-line-number="80"><span class="co">#&gt; C 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb22-81" data-line-number="81"><span class="co">#&gt; G 0 0 0 0 0.0 0 0.0</span></a>
<a class="sourceLine" id="cb22-82" data-line-number="82"><span class="co">#&gt; T 2 0 2 0 0.5 0 0.5</span></a>
<a class="sourceLine" id="cb22-83" data-line-number="83"><span class="co"># convert_type(examplemotif, &quot;ICM&quot;, nsize_correction = TRUE)</span></a>
<a class="sourceLine" id="cb22-84" data-line-number="84">examplemotif[<span class="st">&quot;bkg&quot;</span>] =<span class="st"> </span><span class="kw">c</span>(<span class="fl">0.4</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.4</span>)</a>
<a class="sourceLine" id="cb22-85" data-line-number="85"><span class="kw">convert_type</span>(examplemotif, <span class="st">&quot;ICM&quot;</span>, <span class="dt">relative_entropy =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb22-86" data-line-number="86"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-87" data-line-number="87"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb22-88" data-line-number="88"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb22-89" data-line-number="89"><span class="co">#&gt;              Type:   ICM</span></a>
<a class="sourceLine" id="cb22-90" data-line-number="90"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb22-91" data-line-number="91"><span class="co">#&gt;          Total IC:   14</span></a>
<a class="sourceLine" id="cb22-92" data-line-number="92"><span class="co">#&gt;         Consensus:   TATAWAW</span></a>
<a class="sourceLine" id="cb22-93" data-line-number="93"><span class="co">#&gt;      Target sites:   10</span></a>
<a class="sourceLine" id="cb22-94" data-line-number="94"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb22-95" data-line-number="95"><span class="co">#&gt;      T    A    T    A    W    A    W</span></a>
<a class="sourceLine" id="cb22-96" data-line-number="96"><span class="co">#&gt; A 0.00 1.32 0.00 1.32 0.16 1.32 0.16</span></a>
<a class="sourceLine" id="cb22-97" data-line-number="97"><span class="co">#&gt; C 0.00 0.00 0.00 0.00 0.00 0.00 0.00</span></a>
<a class="sourceLine" id="cb22-98" data-line-number="98"><span class="co">#&gt; G 0.00 0.00 0.00 0.00 0.00 0.00 0.00</span></a>
<a class="sourceLine" id="cb22-99" data-line-number="99"><span class="co">#&gt; T 1.32 0.00 1.32 0.00 0.16 0.00 0.16</span></a></code></pre></div>
</div>
<div id="comparing-and-merging-motifs" class="section level2">
<h2>Comparing and merging motifs</h2>
<p>There a few functions available in other Bioconductor packages which allow for motif comparison. These include <code>PWMSimlarity</code> (<code>Biocpkg(&quot;TFBSTools&quot;)</code>), <code>motifDistances</code> (<em><a href="https://bioconductor.org/packages/3.8/MotIV">MotIV</a></em>), and <code>motifSimilarity</code> (<em><a href="https://bioconductor.org/packages/3.8/PWMEnrich">PWMEnrich</a></em>). Unfortunately these functions are not designed for comparing large numbers of motifs, and can result in long run times. The <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package aims to fix this by providing the <code>compare_motifs</code> function. Using the <code>Biocpkg(&quot;TFBSTools&quot;)</code> implementation as a source, the function has been rewritten in C++ and allows for comparing thousands of motifs in seconds.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="co"># library(MotifDb)</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"></a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="co">## No need to convert class, most universalmotif functions will do it</span></a>
<a class="sourceLine" id="cb23-5" data-line-number="5"><span class="co">## automatically</span></a>
<a class="sourceLine" id="cb23-6" data-line-number="6">motifs.dist &lt;-<span class="st"> </span><span class="kw">compare_motifs</span>(MotifDb[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])</a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="co">#&gt;    0% 50% 75%100%</span></a>
<a class="sourceLine" id="cb23-8" data-line-number="8"><span class="kw">as.dist</span>(motifs.dist)</a>
<a class="sourceLine" id="cb23-9" data-line-number="9"><span class="co">#&gt;                 ABF1       AFT1       AFT2  PK19363.1</span></a>
<a class="sourceLine" id="cb23-10" data-line-number="10"><span class="co">#&gt; AFT1      0.32787228                                 </span></a>
<a class="sourceLine" id="cb23-11" data-line-number="11"><span class="co">#&gt; AFT2      0.52563106 0.66835141                      </span></a>
<a class="sourceLine" id="cb23-12" data-line-number="12"><span class="co">#&gt; PK19363.1 0.09471439 0.39281825 0.51456968           </span></a>
<a class="sourceLine" id="cb23-13" data-line-number="13"><span class="co">#&gt; ORA59     0.01414944 0.46586720 0.45490957 0.90229435</span></a>
<a class="sourceLine" id="cb23-14" data-line-number="14"></a>
<a class="sourceLine" id="cb23-15" data-line-number="15"><span class="co">## If comparing thousands of motifs, multicore usage is made available via</span></a>
<a class="sourceLine" id="cb23-16" data-line-number="16"><span class="co">## the &#39;BiocParallel&#39; package</span></a>
<a class="sourceLine" id="cb23-17" data-line-number="17"></a>
<a class="sourceLine" id="cb23-18" data-line-number="18"><span class="co"># library(BiocParallel)</span></a>
<a class="sourceLine" id="cb23-19" data-line-number="19"><span class="co">## To add a progress bar:</span></a>
<a class="sourceLine" id="cb23-20" data-line-number="20">motifs.dist &lt;-<span class="st"> </span><span class="kw">compare_motifs</span>(MotifDb[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])</a>
<a class="sourceLine" id="cb23-21" data-line-number="21"><span class="co">#&gt;    0% 50% 75%100%</span></a></code></pre></div>
<p>Additionally, <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> provides the <code>merge_motifs</code> function. This first finds the best alignment of input motifs before merging.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="co"># library(MotifDb)</span></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"></a>
<a class="sourceLine" id="cb24-4" data-line-number="4">motifs &lt;-<span class="st"> </span><span class="kw">convert_motifs</span>(MotifDb[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])</a>
<a class="sourceLine" id="cb24-5" data-line-number="5"></a>
<a class="sourceLine" id="cb24-6" data-line-number="6"><span class="co">## Two merging methods are offered: via the &#39;msa&#39; package, and via the </span></a>
<a class="sourceLine" id="cb24-7" data-line-number="7"><span class="co">## &#39;motifStack&#39; package.</span></a>
<a class="sourceLine" id="cb24-8" data-line-number="8"><span class="co"># merge_motifs(motifs, method = &quot;msa&quot;)</span></a>
<a class="sourceLine" id="cb24-9" data-line-number="9"><span class="co"># merge_motifs(motifs, method = &quot;motifStack&quot;)</span></a></code></pre></div>
</div>
</div>
<div id="motif-creation" class="section level1">
<h1>Motif creation</h1>
<p>Though <code>universalmotif</code> class motifs can be created using the <code>new</code> constructor, the <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package provides the <code>create_motif</code> function which aims to provide a simpler interface to motif creation. The <code>universalmotif</code> class was designed to work natively with DNA, RNA, and amino acid motifs. Despite this, it can handle any custom alphabet just as easily.</p>
<p>The <code>create_motif</code> function will be introduced briefly here; see <code>?create_motif</code> for details.</p>
<div id="from-a-pcmppmpwmicm-matrix" class="section level2">
<h2>From a PCM/PPM/PWM/ICM matrix</h2>
<p>Should you wish to make use of the <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> functions starting from a unsupported motif class, you can instead create <code>universalmotif</code> class motifs using the <code>create_motif</code> function.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">motif.matrix &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="fl">0.7</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb25-2" data-line-number="2">                         <span class="fl">0.7</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb25-3" data-line-number="3">                         <span class="fl">0.1</span>, <span class="fl">0.7</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb25-4" data-line-number="4">                         <span class="fl">0.1</span>, <span class="fl">0.7</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb25-5" data-line-number="5">                         <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.7</span>, <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb25-6" data-line-number="6">                         <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.7</span>, <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb25-7" data-line-number="7">                         <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.7</span>,</a>
<a class="sourceLine" id="cb25-8" data-line-number="8">                         <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.7</span>), <span class="dt">nrow =</span> <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb25-9" data-line-number="9">motif &lt;-<span class="st"> </span><span class="kw">create_motif</span>(motif.matrix, <span class="dt">alphabet =</span> <span class="st">&quot;RNA&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;My motif&quot;</span>,</a>
<a class="sourceLine" id="cb25-10" data-line-number="10">                      <span class="dt">pseudocount =</span> <span class="dv">1</span>, <span class="dt">nsites =</span> <span class="dv">20</span>, <span class="dt">strand =</span> <span class="st">&quot;+&quot;</span>)</a>
<a class="sourceLine" id="cb25-11" data-line-number="11"><span class="co">## The &#39;type&#39;, &#39;icscore&#39; and &#39;consensus&#39; slots will be filled for you</span></a>
<a class="sourceLine" id="cb25-12" data-line-number="12">motif</a>
<a class="sourceLine" id="cb25-13" data-line-number="13"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb25-14" data-line-number="14"><span class="co">#&gt;        Motif name:   My motif</span></a>
<a class="sourceLine" id="cb25-15" data-line-number="15"><span class="co">#&gt;          Alphabet:   RNA</span></a>
<a class="sourceLine" id="cb25-16" data-line-number="16"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb25-17" data-line-number="17"><span class="co">#&gt;           Strands:   +</span></a>
<a class="sourceLine" id="cb25-18" data-line-number="18"><span class="co">#&gt;          Total IC:   4.676956</span></a>
<a class="sourceLine" id="cb25-19" data-line-number="19"><span class="co">#&gt;         Consensus:   AACCGGUU</span></a>
<a class="sourceLine" id="cb25-20" data-line-number="20"><span class="co">#&gt;      Target sites:   20</span></a>
<a class="sourceLine" id="cb25-21" data-line-number="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb25-22" data-line-number="22"><span class="co">#&gt;     A   A   C   C   G   G   U   U</span></a>
<a class="sourceLine" id="cb25-23" data-line-number="23"><span class="co">#&gt; A 0.7 0.7 0.1 0.1 0.1 0.1 0.1 0.1</span></a>
<a class="sourceLine" id="cb25-24" data-line-number="24"><span class="co">#&gt; C 0.1 0.1 0.7 0.7 0.1 0.1 0.1 0.1</span></a>
<a class="sourceLine" id="cb25-25" data-line-number="25"><span class="co">#&gt; G 0.1 0.1 0.1 0.1 0.7 0.7 0.1 0.1</span></a>
<a class="sourceLine" id="cb25-26" data-line-number="26"><span class="co">#&gt; U 0.1 0.1 0.1 0.1 0.1 0.1 0.7 0.7</span></a></code></pre></div>
</div>
<div id="from-sequences-or-character-strings" class="section level2">
<h2>From sequences or character strings</h2>
<p>If all you have is a particular consensus sequence in mind, you can easily create a full motif using <code>create_motif</code>. This can be convenient if you’d like to create a motif to use with an external program such as MEME or HOMER.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">motif &lt;-<span class="st"> </span><span class="kw">create_motif</span>(<span class="st">&quot;CCNSNGG&quot;</span>, <span class="dt">nsites =</span> <span class="dv">50</span>, <span class="dt">pseudocount =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="co">## write_meme(motif, &quot;meme_motif.txt&quot;)</span></a>
<a class="sourceLine" id="cb26-3" data-line-number="3"></a>
<a class="sourceLine" id="cb26-4" data-line-number="4">motif</a>
<a class="sourceLine" id="cb26-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb26-6" data-line-number="6"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb26-7" data-line-number="7"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb26-8" data-line-number="8"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb26-9" data-line-number="9"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb26-10" data-line-number="10"><span class="co">#&gt;          Total IC:   8.391796</span></a>
<a class="sourceLine" id="cb26-11" data-line-number="11"><span class="co">#&gt;         Consensus:   CCNSNGG</span></a>
<a class="sourceLine" id="cb26-12" data-line-number="12"><span class="co">#&gt;      Target sites:   50</span></a>
<a class="sourceLine" id="cb26-13" data-line-number="13"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb26-14" data-line-number="14"><span class="co">#&gt;      C    C    N   S    N    G    G</span></a>
<a class="sourceLine" id="cb26-15" data-line-number="15"><span class="co">#&gt; A 0.00 0.00 0.25 0.0 0.25 0.00 0.00</span></a>
<a class="sourceLine" id="cb26-16" data-line-number="16"><span class="co">#&gt; C 0.99 0.99 0.25 0.5 0.25 0.00 0.00</span></a>
<a class="sourceLine" id="cb26-17" data-line-number="17"><span class="co">#&gt; G 0.00 0.00 0.25 0.5 0.25 0.99 0.99</span></a>
<a class="sourceLine" id="cb26-18" data-line-number="18"><span class="co">#&gt; T 0.00 0.00 0.25 0.0 0.25 0.00 0.00</span></a></code></pre></div>
</div>
<div id="generating-random-motifs" class="section level2">
<h2>Generating random motifs</h2>
<p>If you wish, it’s easy to generate random motifs:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">create_motif</span>()</a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb27-3" data-line-number="3"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb27-4" data-line-number="4"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb27-5" data-line-number="5"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb27-6" data-line-number="6"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb27-7" data-line-number="7"><span class="co">#&gt;          Total IC:   11.50575</span></a>
<a class="sourceLine" id="cb27-8" data-line-number="8"><span class="co">#&gt;         Consensus:   CAGARWCCGR</span></a>
<a class="sourceLine" id="cb27-9" data-line-number="9"><span class="co">#&gt;      Target sites:   143</span></a>
<a class="sourceLine" id="cb27-10" data-line-number="10"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb27-11" data-line-number="11"><span class="co">#&gt;      C    A    G    A    R    W   C C    G    R</span></a>
<a class="sourceLine" id="cb27-12" data-line-number="12"><span class="co">#&gt; A 0.05 0.62 0.05 0.86 0.49 0.31 0.0 0 0.07 0.48</span></a>
<a class="sourceLine" id="cb27-13" data-line-number="13"><span class="co">#&gt; C 0.95 0.25 0.23 0.08 0.00 0.06 0.8 1 0.00 0.00</span></a>
<a class="sourceLine" id="cb27-14" data-line-number="14"><span class="co">#&gt; G 0.01 0.00 0.72 0.05 0.33 0.05 0.2 0 0.93 0.52</span></a>
<a class="sourceLine" id="cb27-15" data-line-number="15"><span class="co">#&gt; T 0.00 0.13 0.00 0.02 0.18 0.58 0.0 0 0.00 0.00</span></a>
<a class="sourceLine" id="cb27-16" data-line-number="16"></a>
<a class="sourceLine" id="cb27-17" data-line-number="17"><span class="co">## You can change the probabilities:</span></a>
<a class="sourceLine" id="cb27-18" data-line-number="18"><span class="kw">create_motif</span>(<span class="dt">bkg =</span> <span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.4</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>))</a>
<a class="sourceLine" id="cb27-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb27-20" data-line-number="20"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb27-21" data-line-number="21"><span class="co">#&gt;          Alphabet:   DNA</span></a>
<a class="sourceLine" id="cb27-22" data-line-number="22"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb27-23" data-line-number="23"><span class="co">#&gt;           Strands:   +-</span></a>
<a class="sourceLine" id="cb27-24" data-line-number="24"><span class="co">#&gt;          Total IC:   11.32007</span></a>
<a class="sourceLine" id="cb27-25" data-line-number="25"><span class="co">#&gt;         Consensus:   AWCATCTCCC</span></a>
<a class="sourceLine" id="cb27-26" data-line-number="26"><span class="co">#&gt;      Target sites:   182</span></a>
<a class="sourceLine" id="cb27-27" data-line-number="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb27-28" data-line-number="28"><span class="co">#&gt;      A    W    C    A    T    C    T    C    C    C</span></a>
<a class="sourceLine" id="cb27-29" data-line-number="29"><span class="co">#&gt; A 0.91 0.56 0.02 0.66 0.30 0.17 0.14 0.00 0.17 0.02</span></a>
<a class="sourceLine" id="cb27-30" data-line-number="30"><span class="co">#&gt; C 0.08 0.00 0.95 0.08 0.05 0.75 0.00 0.86 0.58 0.93</span></a>
<a class="sourceLine" id="cb27-31" data-line-number="31"><span class="co">#&gt; G 0.00 0.00 0.03 0.26 0.03 0.08 0.00 0.08 0.21 0.05</span></a>
<a class="sourceLine" id="cb27-32" data-line-number="32"><span class="co">#&gt; T 0.01 0.44 0.00 0.00 0.62 0.00 0.86 0.06 0.04 0.00</span></a>
<a class="sourceLine" id="cb27-33" data-line-number="33"></a>
<a class="sourceLine" id="cb27-34" data-line-number="34"><span class="co">## With a custom alphabet:</span></a>
<a class="sourceLine" id="cb27-35" data-line-number="35"><span class="kw">create_motif</span>(<span class="dt">alphabet =</span> <span class="st">&quot;QWERTY&quot;</span>)</a>
<a class="sourceLine" id="cb27-36" data-line-number="36"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb27-37" data-line-number="37"><span class="co">#&gt;        Motif name:   motif</span></a>
<a class="sourceLine" id="cb27-38" data-line-number="38"><span class="co">#&gt;          Alphabet:   QWERTY</span></a>
<a class="sourceLine" id="cb27-39" data-line-number="39"><span class="co">#&gt;              Type:   PPM</span></a>
<a class="sourceLine" id="cb27-40" data-line-number="40"><span class="co">#&gt;          Total IC:   15.2096</span></a>
<a class="sourceLine" id="cb27-41" data-line-number="41"><span class="co">#&gt;      Target sites:   159</span></a>
<a class="sourceLine" id="cb27-42" data-line-number="42"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb27-43" data-line-number="43"><span class="co">#&gt;   [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></a>
<a class="sourceLine" id="cb27-44" data-line-number="44"><span class="co">#&gt; E 0.00 0.00 0.09 0.74 0.03 0.00 0.04 0.00 0.00  0.00</span></a>
<a class="sourceLine" id="cb27-45" data-line-number="45"><span class="co">#&gt; Q 0.03 0.00 0.00 0.18 0.43 0.00 0.00 0.00 0.06  0.73</span></a>
<a class="sourceLine" id="cb27-46" data-line-number="46"><span class="co">#&gt; R 0.06 0.00 0.00 0.00 0.00 0.01 0.02 0.11 0.23  0.00</span></a>
<a class="sourceLine" id="cb27-47" data-line-number="47"><span class="co">#&gt; T 0.44 0.43 0.90 0.00 0.00 0.43 0.00 0.16 0.00  0.00</span></a>
<a class="sourceLine" id="cb27-48" data-line-number="48"><span class="co">#&gt; W 0.47 0.01 0.00 0.00 0.54 0.28 0.00 0.00 0.08  0.00</span></a>
<a class="sourceLine" id="cb27-49" data-line-number="49"><span class="co">#&gt; Y 0.00 0.56 0.00 0.08 0.00 0.28 0.95 0.72 0.62  0.27</span></a></code></pre></div>
</div>
</div>
<div id="motif-visualization" class="section level1">
<h1>Motif visualization</h1>
<div id="motif-logos" class="section level2">
<h2>Motif logos</h2>
<p>There are several packages which offer motif visualization capabilities, such as <em><a href="https://bioconductor.org/packages/3.8/seqLogo">seqLogo</a></em>, <em><a href="https://bioconductor.org/packages/3.8/Logolas">Logolas</a></em>, <code>Biocpkg(&quot;motifStack&quot;)</code>, and <em><a href="https://CRAN.R-project.org/package=ggseqlogo">ggseqlogo</a></em>. Here I will briefly show how to use these to visualize <code>universalmotif</code> class motifs.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="co"># library(universalmotif)</span></a>
<a class="sourceLine" id="cb28-2" data-line-number="2"><span class="kw">data</span>(examplemotif)</a>
<a class="sourceLine" id="cb28-3" data-line-number="3"></a>
<a class="sourceLine" id="cb28-4" data-line-number="4"><span class="co">## For all the examples, simply passing the functions a PPM is sufficient</span></a>
<a class="sourceLine" id="cb28-5" data-line-number="5">motif &lt;-<span class="st"> </span><span class="kw">convert_type</span>(examplemotif, <span class="st">&quot;PPM&quot;</span>)</a>
<a class="sourceLine" id="cb28-6" data-line-number="6"><span class="co">## Only need the matrix itself</span></a>
<a class="sourceLine" id="cb28-7" data-line-number="7">motif &lt;-<span class="st"> </span>motif[<span class="st">&quot;motif&quot;</span>]</a>
<a class="sourceLine" id="cb28-8" data-line-number="8"></a>
<a class="sourceLine" id="cb28-9" data-line-number="9"><span class="co">## seqLogo:</span></a>
<a class="sourceLine" id="cb28-10" data-line-number="10">seqLogo<span class="op">::</span><span class="kw">seqLogo</span>(motif)</a></code></pre></div>
<p><img src="MotifManipulation_files/figure-html/unnamed-chunk-11-1.png" width="700" /></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"></a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="co">## motifStack: </span></a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="co"># motifStack::plotMotifLogo(motif)</span></a>
<a class="sourceLine" id="cb29-4" data-line-number="4"></a>
<a class="sourceLine" id="cb29-5" data-line-number="5"><span class="co">## Logolas:</span></a>
<a class="sourceLine" id="cb29-6" data-line-number="6"><span class="cf">if</span> (<span class="kw">requireNamespace</span>(<span class="st">&quot;Logolas&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {</a>
<a class="sourceLine" id="cb29-7" data-line-number="7">  <span class="kw">colnames</span>(motif) &lt;-<span class="st"> </span><span class="kw">seq_len</span>(<span class="kw">ncol</span>(motif))</a>
<a class="sourceLine" id="cb29-8" data-line-number="8">  Logolas<span class="op">::</span><span class="kw">logomaker</span>(motif, <span class="dt">type =</span> <span class="st">&quot;Logo&quot;</span>)</a>
<a class="sourceLine" id="cb29-9" data-line-number="9">}</a>
<a class="sourceLine" id="cb29-10" data-line-number="10"><span class="co">#&gt; color_type not provided, so switching to per_row option for</span></a>
<a class="sourceLine" id="cb29-11" data-line-number="11"><span class="co">#&gt;                color_type</span></a>
<a class="sourceLine" id="cb29-12" data-line-number="12"><span class="co">#&gt; frame width not provided, taken to be 1</span></a>
<a class="sourceLine" id="cb29-13" data-line-number="13"><span class="co">#&gt; using a background with equal probability for all symbols</span></a></code></pre></div>
<p><img src="MotifManipulation_files/figure-html/unnamed-chunk-11-2.png" width="700" /></p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"></a>
<a class="sourceLine" id="cb30-2" data-line-number="2"><span class="co">## ggseqlogo:</span></a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="cf">if</span> (<span class="kw">requireNamespace</span>(<span class="st">&quot;ggseqlogo&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {</a>
<a class="sourceLine" id="cb30-4" data-line-number="4">  ggseqlogo<span class="op">::</span><span class="kw">ggseqlogo</span>(motif)</a>
<a class="sourceLine" id="cb30-5" data-line-number="5">}</a></code></pre></div>
<p><img src="MotifManipulation_files/figure-html/unnamed-chunk-11-3.png" width="700" /></p>
<p>The <em><a href="https://bioconductor.org/packages/3.8/Logolas">Logolas</a></em> and <em><a href="https://CRAN.R-project.org/package=ggseqlogo">ggseqlogo</a></em> offer many additional options for logo customization, including custom alphabets as well as manually determining the heights of each letter, via the <em>grid</em> and <em><a href="https://CRAN.R-project.org/package=ggplot2">ggplot2</a></em> packages respectively.</p>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1>Session info</h1>
<pre><code>#&gt; R version 3.5.1 (2018-07-02)
#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)
#&gt; Running under: macOS High Sierra 10.13.4
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
#&gt; [8] methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] MotifDb_1.23.0        Biostrings_2.49.1     XVector_0.21.3       
#&gt; [4] IRanges_2.15.17       S4Vectors_0.19.19     BiocGenerics_0.27.1  
#&gt; [7] universalmotif_0.99.0 BiocStyle_2.9.6       nvimcom_0.9-75       
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] Biobase_2.41.2              tidyr_0.8.1                
#&gt;  [3] jsonlite_1.5                gtools_3.8.1               
#&gt;  [5] Rdpack_0.9-1                assertthat_0.2.0           
#&gt;  [7] BiocManager_1.30.2          rvcheck_0.1.0              
#&gt;  [9] GenomeInfoDbData_1.1.0      Rsamtools_1.33.5           
#&gt; [11] yaml_2.2.0                  pillar_1.3.0               
#&gt; [13] backports_1.1.2             lattice_0.20-35            
#&gt; [15] glue_1.3.0                  digest_0.6.17              
#&gt; [17] RColorBrewer_1.1-2          GenomicRanges_1.33.13      
#&gt; [19] colorspace_1.3-2            Matrix_1.2-14              
#&gt; [21] htmltools_0.3.6             plyr_1.8.4                 
#&gt; [23] XML_3.98-1.16               pkgconfig_2.0.2            
#&gt; [25] bibtex_0.4.2                ggseqlogo_0.1              
#&gt; [27] bookdown_0.7                zlibbioc_1.27.0            
#&gt; [29] purrr_0.2.5                 tidytree_0.1.9             
#&gt; [31] scales_1.0.0                processx_3.2.0             
#&gt; [33] BiocParallel_1.15.12        tibble_1.4.2               
#&gt; [35] ggplot2_3.0.0               SummarizedExperiment_1.11.6
#&gt; [37] lazyeval_0.2.1              splitstackshape_1.4.6      
#&gt; [39] magrittr_1.5                crayon_1.3.4               
#&gt; [41] memoise_1.1.0               evaluate_0.11              
#&gt; [43] ps_1.1.0                    fs_1.2.6                   
#&gt; [45] nlme_3.1-137                MASS_7.3-50                
#&gt; [47] xml2_1.2.0                  data.table_1.11.4          
#&gt; [49] tools_3.5.1                 gridBase_0.4-7             
#&gt; [51] matrixStats_0.54.0          gbRd_0.4-11                
#&gt; [53] stringr_1.3.1               ggtree_1.13.3              
#&gt; [55] munsell_0.5.0               DelayedArray_0.7.41        
#&gt; [57] bindrcpp_0.2.2              compiler_3.5.1             
#&gt; [59] pkgdown_1.1.0.9000          GenomeInfoDb_1.17.1        
#&gt; [61] rlang_0.2.2                 debugme_1.1.0              
#&gt; [63] grid_3.5.1                  RCurl_1.95-4.11            
#&gt; [65] labeling_0.3                bitops_1.0-6               
#&gt; [67] rmarkdown_1.10              Logolas_1.5.1              
#&gt; [69] gtable_0.2.0                roxygen2_6.1.0             
#&gt; [71] R6_2.2.2                    GenomicAlignments_1.17.3   
#&gt; [73] seqLogo_1.47.0              knitr_1.20                 
#&gt; [75] dplyr_0.7.6                 rtracklayer_1.41.5         
#&gt; [77] bindr_0.1.1                 commonmark_1.6             
#&gt; [79] rprojroot_1.3-2             treeio_1.5.5               
#&gt; [81] desc_1.2.0                  ape_5.1                    
#&gt; [83] stringi_1.2.4               SQUAREM_2017.10-1          
#&gt; [85] Rcpp_0.12.18                tidyselect_0.2.4           
#&gt; [87] xfun_0.3</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-meme">
<p>Bailey, T.L., M. Boden, F.A. Buske, M. Frith, C.E. Grant, L. Clementi, J. Ren, W.W. Li, and W.S. Noble. 2009. “MEME Suite: Tools for Motif Discovery and Searching.” <em>Nucleic Acids Research</em> 37: W202–W208.</p>
</div>
<div id="ref-homer">
<p>Heinz, S., C. Benner, N. Spann, E. Bertolino, Y.C. Lin, P. Laslo, J.X. Cheng, C. Murre, H. Singh, and C.K. Glass. 2010. “Simple Combinations of Lineage-Determining Transcription Factors Prime Cis-Regulatory Elements Required for Macrophage and B Cell Identities.” <em>Molecular Cell</em> 38 (4): 576–89.</p>
</div>
<div id="ref-uniprobe">
<p>Hume, M.A., L.A. Barrera, S.S. Gisselbrecht, and M.L. Bulyk. 2015. “UniPROBE, Update 2015: New Tools and Content for the Online Database of Protein-Binding Microarray Data on Protein-Dna Interactions.” <em>Nucleic Acids Research</em> 43: D117–D122.</p>
</div>
<div id="ref-jaspar">
<p>Khan, A., O. Fornes, A. Stigliani, M. Gheorghe, J.A. Castro-Mondragon, R. van der Lee, A. Bessy, et al. 2018. “JASPAR 2018: Update of the Open-Access Database of Transcription Factor Binding Profiles and Its Web Framework.” <em>Nucleic Acids Research</em> 46 (D1): D260–D266.</p>
</div>
<div id="ref-cisbp">
<p>Weirauch, M.T., A. Yang, M. Albu, A.G. Cote, A. Montenegro-Montero, P. Drewe, H.S. Najafabadi, et al. 2014. “Determination and Inference of Eukaryotic Transcription Factor Sequence Specificity.” <em>Cell</em> 158 (6): 1431–43.</p>
</div>
<div id="ref-transfac">
<p>Wingender, E., P. Dietze, H. Karas, and R. Knuppel. 1996. “TRANSFAC: A Database on Transcription Factors and Their Dna Binding Sites.” <em>Nucleic Acids Research</em> 24 (1): 238–41.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul>
      <li><a href="#the-universalmotif-class-and-conversion-utilities">The universalmotif class and conversion utilities</a><ul>
      <li><a href="#the-universalmotif-class">The universalmotif class</a></li>
      <li><a href="#converting-to-and-from-another-packages-class">Converting to and from another package’s class</a></li>
      </ul></li>
      <li><a href="#importing-and-exporting-motifs">Importing and exporting motifs</a><ul>
      <li><a href="#importing">Importing</a></li>
      <li><a href="#exporting">Exporting</a></li>
      </ul></li>
      <li><a href="#modifying-motifs-and-related-functions">Modifying motifs and related functions</a><ul>
      <li><a href="#converting-motif-type">Converting motif type</a></li>
      <li><a href="#comparing-and-merging-motifs">Comparing and merging motifs</a></li>
      </ul></li>
      <li><a href="#motif-creation">Motif creation</a><ul>
      <li><a href="#from-a-pcmppmpwmicm-matrix">From a PCM/PPM/PWM/ICM matrix</a></li>
      <li><a href="#from-sequences-or-character-strings">From sequences or character strings</a></li>
      <li><a href="#generating-random-motifs">Generating random motifs</a></li>
      </ul></li>
      <li><a href="#motif-visualization">Motif visualization</a><ul>
      <li><a href="#motif-logos">Motif logos</a></li>
      </ul></li>
      <li><a href="#session-info">Session info</a></li>
      <li><a href="#references">References</a></li>
      </ul>
    </div>
      </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Benjamin Tremblay.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>
