<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sequence creation, manipulation, and scanning â€¢ universalmotif</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Sequence creation, manipulation, and scanning">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">universalmotif</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AdvancedUsage.html">Advanced motif and sequence manipulation</a>
    </li>
    <li>
      <a href="../articles/Introduction.html">Introduction to the "universalmotif" package</a>
    </li>
    <li>
      <a href="../articles/IntroductionToSequenceMotifs.html">Introduction to sequence motifs</a>
    </li>
    <li>
      <a href="../articles/MotifManipulation.html">Motif import, export, visualization, and manipulation</a>
    </li>
    <li>
      <a href="../articles/SequenceSearches.html">Sequence creation, manipulation, and scanning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bjmt/universalmotif">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Sequence creation, manipulation, and scanning</h1>
                        <h4 class="author">Benjamin Tremblay</h4>
            <address class="author_afil">
      University of Waterloo, Waterloo, Canada<br><a class="author_email" href="mailto:#"></a><a href="mailto:b2tremblay@uwaterloo.ca">b2tremblay@uwaterloo.ca</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/bjmt/universalmotif/blob/master/R/vignettes/SequenceSearches.Rmd"><code>R/vignettes/SequenceSearches.Rmd</code></a></small>
      <div class="hidden name"><code>SequenceSearches.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Introducing sequence motif types.
    </div>
    
<div id="sequence-related-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#sequence-related-functions" class="anchor"></a>Sequence-related functions</h1>
<div id="sequence-shuffling-and-creation" class="section level2">
<h2 class="hasAnchor">
<a href="#sequence-shuffling-and-creation" class="anchor"></a>Sequence shuffling and creation</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.8/Biostrings">Biostrings</a></em> package offers an excellent suite of functions for dealing with biological sequences. The <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package hopes to help extend these by providing the <code>create_sequences</code> and <code>shuffle_sequences</code> functions.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(universalmotif)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">## library(Biostrings)</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">## Create some DNA sequences for use with an external program:</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">sequences &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_sequences.html">create_sequences</a></span>(<span class="dt">seqnum =</span> <span class="dv">500</span>, <span class="dt">monofreqs =</span> <span class="kw">c</span>(<span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>))</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">## writeXStringSet(sequences, "bkg.fasta")</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">## Shuffle sequences:</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">## sequences &lt;- readDNAStringSet("sequences.fasta")</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9">sequences &lt;-<span class="st"> </span><span class="kw"><a href="../reference/shuffle_sequences.html">shuffle_sequences</a></span>(sequences)</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co">## Maintain dinucleotide frequencies:</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12">sequences &lt;-<span class="st"> </span><span class="kw"><a href="../reference/shuffle_sequences.html">shuffle_sequences</a></span>(sequences, <span class="dt">k =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb1-13" data-line-number="13"></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co">## Create sequences from any starting alphabet:</span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15">sequences &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_sequences.html">create_sequences</a></span>(<span class="dt">alphabet =</span> <span class="st">"QWERTY"</span>)</a>
<a class="sourceLine" id="cb1-16" data-line-number="16">sequences &lt;-<span class="st"> </span><span class="kw"><a href="../reference/shuffle_sequences.html">shuffle_sequences</a></span>(sequences, <span class="dt">k =</span> <span class="dv">6</span>)</a></code></pre></div>
</div>
<div id="searching-for-motifs-in-sequences" class="section level2">
<h2 class="hasAnchor">
<a href="#searching-for-motifs-in-sequences" class="anchor"></a>Searching for motifs in sequences</h2>
<p>There are many fantastic motif scanning and enrichment programs available, such as <a href="http://homer.ucsd.edu/homer/index.html">HOMER</a> and the <a href="http://meme-suite.org/">MEME suite</a> to name a couple. The <em><a href="https://bioconductor.org/packages/3.8/universalmotif">universalmotif</a></em> package does not aim to supplant these, but rather provide convenience functions for quickly scanning a few sequences without needing to leave the R environment. Furthermore, these functions allow for taking advantage of the k-letter motif format described earlier.</p>
<p>Two functions are provided: <code>scan_sequences</code> and <code>enrich_motifs</code>. Given a motif of length n, <code>scan_sequences</code> considers every n-length subset in a sequence and scores it using the PWM format. If it surpasses the minimum threshold, it is reported. This is case regardless of whether one is scanning with a regular motif, or using the k-letter motif format (the k-letter PPM is converted to a k-letter PWM). The second function, <code>enrich_motifs</code>, simply scans a set of target sequences, then a set of background sequences, and checks for significant enrichment.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(universalmotif)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw">library</span>(Biostrings)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co">#&gt; Loading required package: BiocGenerics</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#&gt; Loading required package: parallel</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; The following objects are masked from 'package:parallel':</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind,</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co">#&gt;     colMeans, colnames, colSums, dirname, do.call, duplicated,</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co">#&gt;     eval, evalq, Filter, Find, get, grep, grepl, intersect,</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">#&gt;     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,</span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">#&gt;     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co">#&gt;     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,</span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co">#&gt;     table, tapply, union, unique, unsplit, which, which.max,</span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co">#&gt;     which.min</span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25"><span class="co">#&gt; Loading required package: S4Vectors</span></a>
<a class="sourceLine" id="cb2-26" data-line-number="26"><span class="co">#&gt; Loading required package: stats4</span></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="co">#&gt; Attaching package: 'S4Vectors'</span></a>
<a class="sourceLine" id="cb2-29" data-line-number="29"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb2-30" data-line-number="30"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-31" data-line-number="31"><span class="co">#&gt;     expand.grid</span></a>
<a class="sourceLine" id="cb2-32" data-line-number="32"><span class="co">#&gt; Loading required package: IRanges</span></a>
<a class="sourceLine" id="cb2-33" data-line-number="33"><span class="co">#&gt; Loading required package: XVector</span></a>
<a class="sourceLine" id="cb2-34" data-line-number="34"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-35" data-line-number="35"><span class="co">#&gt; Attaching package: 'Biostrings'</span></a>
<a class="sourceLine" id="cb2-36" data-line-number="36"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb2-37" data-line-number="37"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb2-38" data-line-number="38"><span class="co">#&gt;     strsplit</span></a>
<a class="sourceLine" id="cb2-39" data-line-number="39"></a>
<a class="sourceLine" id="cb2-40" data-line-number="40">sequences &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_sequences.html">create_sequences</a></span>()</a>
<a class="sourceLine" id="cb2-41" data-line-number="41"></a>
<a class="sourceLine" id="cb2-42" data-line-number="42"><span class="co">## Let us revisit our 2-letter motif example:</span></a>
<a class="sourceLine" id="cb2-43" data-line-number="43">motif.k2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_motif.html">create_motif</a></span>(<span class="st">"CWWWWCC"</span>, <span class="dt">nsites =</span> <span class="dv">6</span>)</a>
<a class="sourceLine" id="cb2-44" data-line-number="44">sequences.k2 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/Biostrings/topics/XStringSet-class">DNAStringSet</a></span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">"CAAAACC"</span>, <span class="st">"CTTTTCC"</span>), <span class="dv">3</span>))</a>
<a class="sourceLine" id="cb2-45" data-line-number="45">motif.k2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/add_multifreq.html">add_multifreq</a></span>(motif.k2, sequences.k2)</a>
<a class="sourceLine" id="cb2-46" data-line-number="46"></a>
<a class="sourceLine" id="cb2-47" data-line-number="47"><span class="co">## Regular scanning:</span></a>
<a class="sourceLine" id="cb2-48" data-line-number="48"><span class="kw">head</span>(<span class="kw"><a href="../reference/scan_sequences.html">scan_sequences</a></span>(motif.k2, sequences, <span class="dt">RC =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb2-49" data-line-number="49">                    <span class="dt">threshold =</span> <span class="fl">0.0001</span>))</a>
<a class="sourceLine" id="cb2-50" data-line-number="50"><span class="co">#&gt; Warning in motif_pvalue(motifs, pvalue = threshold, use.freq = use.freq, :</span></a>
<a class="sourceLine" id="cb2-51" data-line-number="51"><span class="co">#&gt; Found -Inf values in motif PWM, adding a pseudocount of 1</span></a>
<a class="sourceLine" id="cb2-52" data-line-number="52"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-53" data-line-number="53"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-54" data-line-number="54"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-55" data-line-number="55"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-56" data-line-number="56"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-57" data-line-number="57"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-58" data-line-number="58"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-59" data-line-number="59"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-60" data-line-number="60"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-61" data-line-number="61"><span class="co">#&gt;   motif sequence start stop score max.score score.pct   match strand</span></a>
<a class="sourceLine" id="cb2-62" data-line-number="62"><span class="co">#&gt; 1 motif        6    19   25    10        10       100 CAAAACC      +</span></a>
<a class="sourceLine" id="cb2-63" data-line-number="63"><span class="co">#&gt; 2 motif       13     2    8    10        10       100 CTATTCC      +</span></a>
<a class="sourceLine" id="cb2-64" data-line-number="64"><span class="co">#&gt; 3 motif       26    17   23    10        10       100 CAATTCC      +</span></a>
<a class="sourceLine" id="cb2-65" data-line-number="65"><span class="co">#&gt; 4 motif       35    79   85    10        10       100 CTATTCC      +</span></a>
<a class="sourceLine" id="cb2-66" data-line-number="66"><span class="co">#&gt; 5 motif       40    67   73    10        10       100 CATAACC      +</span></a>
<a class="sourceLine" id="cb2-67" data-line-number="67"><span class="co">#&gt; 6 motif       43     6   12    10        10       100 CTATTCC      +</span></a>
<a class="sourceLine" id="cb2-68" data-line-number="68"></a>
<a class="sourceLine" id="cb2-69" data-line-number="69"><span class="co">## Using 2-letter information to scan:</span></a>
<a class="sourceLine" id="cb2-70" data-line-number="70"><span class="kw">head</span>(<span class="kw"><a href="../reference/scan_sequences.html">scan_sequences</a></span>(motif.k2, sequences, <span class="dt">use.freq =</span> <span class="dv">2</span>, <span class="dt">RC =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb2-71" data-line-number="71"><span class="co">#&gt; Warning in motif_pvalue(motifs, pvalue = threshold, use.freq = use.freq, :</span></a>
<a class="sourceLine" id="cb2-72" data-line-number="72"><span class="co">#&gt; Found -Inf values in motif PWM, adding a pseudocount of 1</span></a>
<a class="sourceLine" id="cb2-73" data-line-number="73"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-74" data-line-number="74"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-75" data-line-number="75"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-76" data-line-number="76"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-77" data-line-number="77"><span class="co">#&gt;    0%  2%  3%  4%  5%  6%  7%  8%  9% 10% 11% 12% 13% 14% 15% 16% 17% 18% 19% 20% 21% 22% 23% 24% 25% 26% 27% 28% 29% 30% 31% 32% 33% 34% 35% 36% 37% 38% 39% 40% 41% 42% 43% 44% 45% 46% 47% 48% 49% 50% 51% 52% 53% 54% 55% 56% 57% 58% 59% 60% 61% 62% 63% 64% 65% 66% 67% 68% 69% 70% 71% 72% 73% 74% 75% 76% 77% 78% 79% 80% 81% 82% 83% 84% 85% 86% 87% 88% 89% 90% 91% 92% 93% 94% 95% 96% 97% 98% 99%100%</span></a>
<a class="sourceLine" id="cb2-78" data-line-number="78"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-79" data-line-number="79"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-80" data-line-number="80"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-81" data-line-number="81"><span class="co">#&gt;    0%100%</span></a>
<a class="sourceLine" id="cb2-82" data-line-number="82"><span class="co">#&gt;   motif sequence start stop  score max.score score.pct   match strand</span></a>
<a class="sourceLine" id="cb2-83" data-line-number="83"><span class="co">#&gt; 1 motif        6    19   26 18.994  18.99872  99.97514 CAAAACC      +</span></a>
<a class="sourceLine" id="cb2-84" data-line-number="84"><span class="co">#&gt; 2 motif       95    27   34 18.994  18.99872  99.97514 CAAAACC      +</span></a></code></pre></div>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>#&gt; R version 3.5.1 (2018-07-02)
#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)
#&gt; Running under: macOS High Sierra 10.13.4
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
#&gt; [8] methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] Biostrings_2.49.1     XVector_0.21.3        IRanges_2.15.17      
#&gt; [4] S4Vectors_0.19.19     BiocGenerics_0.27.1   universalmotif_0.99.0
#&gt; [7] BiocStyle_2.9.6       nvimcom_0.9-75       
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] Rcpp_0.12.18       ape_5.1            lattice_0.20-35   
#&gt;  [4] tidyr_0.8.1        ps_1.1.0           ggseqlogo_0.1     
#&gt;  [7] gtools_3.8.1       assertthat_0.2.0   rprojroot_1.3-2   
#&gt; [10] digest_0.6.17      R6_2.2.2           plyr_1.8.4        
#&gt; [13] backports_1.1.2    evaluate_0.11      ggplot2_3.0.0     
#&gt; [16] pillar_1.3.0       Rdpack_0.9-1       zlibbioc_1.27.0   
#&gt; [19] rlang_0.2.2        lazyeval_0.2.1     rmarkdown_1.10    
#&gt; [22] pkgdown_1.1.0.9000 desc_1.2.0         stringr_1.3.1     
#&gt; [25] munsell_0.5.0      compiler_3.5.1     xfun_0.3          
#&gt; [28] pkgconfig_2.0.2    htmltools_0.3.6    tidyselect_0.2.4  
#&gt; [31] tibble_1.4.2       roxygen2_6.1.0     bookdown_0.7      
#&gt; [34] crayon_1.3.4       dplyr_0.7.6        MASS_7.3-50       
#&gt; [37] commonmark_1.6     grid_3.5.1         nlme_3.1-137      
#&gt; [40] jsonlite_1.5       gtable_0.2.0       magrittr_1.5      
#&gt; [43] scales_1.0.0       bibtex_0.4.2       tidytree_0.1.9    
#&gt; [46] stringi_1.2.4      debugme_1.1.0      fs_1.2.6          
#&gt; [49] bindrcpp_0.2.2     ggtree_1.13.3      xml2_1.2.0        
#&gt; [52] rvcheck_0.1.0      tools_3.5.1        treeio_1.5.5      
#&gt; [55] glue_1.3.0         purrr_0.2.5        processx_3.2.0    
#&gt; [58] yaml_2.2.0         colorspace_1.3-2   BiocManager_1.30.2
#&gt; [61] gbRd_0.4-11        memoise_1.1.0      knitr_1.20        
#&gt; [64] bindr_0.1.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#sequence-related-functions">Sequence-related functions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#sequence-shuffling-and-creation">Sequence shuffling and creation</a></li>
      <li><a href="#searching-for-motifs-in-sequences">Searching for motifs in sequences</a></li>
      </ul>
</li>
      <li><a href="#session-info">Session info</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Benjamin Tremblay.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
