<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compare motifs. — compare_motifs • universalmotif</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Compare motifs. — compare_motifs" />

<meta property="og:description" content="Compare motifs using four available metrics: Pearson correlation coefficient
(Pietrokovski 1996)
, Euclidean distance
(Choi et al. 2004)
, Sandelin-Wasserman similarity
(Sandelin and Wasserman 2004)
, and Kullback-Leibler divergence
(Roepcke et al. 2005)
." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">universalmotif</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/AdvancedUsage.html">Advanced motif and sequence manipulation</a>
    </li>
    <li>
      <a href="../articles/Introduction.html">Introduction to the "universalmotif" package</a>
    </li>
    <li>
      <a href="../articles/IntroductionToSequenceMotifs.html">Introduction to sequence motifs</a>
    </li>
    <li>
      <a href="../articles/MotifManipulation.html">Motif import, export, visualization, and manipulation</a>
    </li>
    <li>
      <a href="../articles/SequenceSearches.html">Sequence creation, manipulation, and scanning</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bjmt/universalmotif">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compare motifs.</h1>
    <small class="dont-index">Source: <a href='https://github.com/bjmt/universalmotif/blob/master/R/compare_motifs.R'><code>R/compare_motifs.R</code></a></small>
    <div class="hidden name"><code>compare_motifs.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Compare motifs using four available metrics: Pearson correlation coefficient
(Pietrokovski 1996)
, Euclidean distance
(Choi et al. 2004)
, Sandelin-Wasserman similarity
(Sandelin and Wasserman 2004)
, and Kullback-Leibler divergence
(Roepcke et al. 2005)
.</p>
    
    </div>

    <pre class="usage"><span class='fu'>compare_motifs</span>(<span class='no'>motifs</span>, <span class='no'>compare.to</span>, <span class='no'>db.scores</span>, <span class='kw'>use.freq</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>use.type</span> <span class='kw'>=</span> <span class='st'>"PPM"</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"MPCC"</span>, <span class='kw'>tryRC</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>min.overlap</span> <span class='kw'>=</span> <span class='fl'>6</span>,
  <span class='kw'>min.mean.ic</span> <span class='kw'>=</span> <span class='fl'>0.5</span>, <span class='kw'>relative_entropy</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>normalise.scores</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>max.p</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>max.e</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>progress</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>BP</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>motifs</th>
      <td><p>See <code><a href='convert_motifs.html'>convert_motifs()</a></code> for acceptable motif formats.</p></td>
    </tr>
    <tr>
      <th>compare.to</th>
      <td><p><code>numeric</code> If missing, compares all motifs to all other motifs.
Otherwise compares all motifs to the specified motif(s).</p></td>
    </tr>
    <tr>
      <th>db.scores</th>
      <td><p><code>data.frame</code> See <code>details</code>.</p></td>
    </tr>
    <tr>
      <th>use.freq</th>
      <td><p><code>numeric(1)</code>. For comparing the <code>multifreq</code> slot.</p></td>
    </tr>
    <tr>
      <th>use.type</th>
      <td><p><code>character(1)</code> One of <code>'PPM'</code> and <code>'ICM'</code>.
The latter allows for taking into account the background
frequencies if <code>relative_entropy = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p><code>character(1)</code> One of <code>c('PCC', 'MPCC', 'EUCL', 'MEUCL', 'SW', 'MSW', 'KL', 'MKL')</code>. See details.</p></td>
    </tr>
    <tr>
      <th>tryRC</th>
      <td><p><code>logical</code> Try the reverse complement of the motifs as well,
report the best score.</p></td>
    </tr>
    <tr>
      <th>min.overlap</th>
      <td><p><code>numeric(1)</code> Minimum overlap required when aligning the
motifs. Setting this to a number higher then the width of the motifs
will not allow any overhangs. Can also be a number less than 1,
representing the minimum fraction that the motifs must overlap.</p></td>
    </tr>
    <tr>
      <th>min.mean.ic</th>
      <td><p><code>numeric(1)</code> Minimum information content between the
two motifs for an alignment to be scored. This helps prevent scoring
alignments between low information content regions of two motifs.</p></td>
    </tr>
    <tr>
      <th>relative_entropy</th>
      <td><p><code>logical(1)</code> For ICM calculation. See
<code><a href='convert_type.html'>convert_type()</a></code>.</p></td>
    </tr>
    <tr>
      <th>normalise.scores</th>
      <td><p><code>logical(1)</code> Favour alignments which leave fewer
unaligned positions, as well as alignments between motifs of similar length.
Similarity scores are multiplied by the ratio of
aligned positions to the total number of positions in the larger motif,
and the inverse for distance scores.</p></td>
    </tr>
    <tr>
      <th>max.p</th>
      <td><p><code>numeric(1)</code> Maximum P-value allowed in reporting matches.
Only used if <code>compare.to</code> is set.</p></td>
    </tr>
    <tr>
      <th>max.e</th>
      <td><p><code>numeric(1)</code> Maximum E-value allowed in reporting matches.
Only used if <code>compare.to</code> is set. The E-value is the P-value multiplied
by the number of input motifs times two.</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p><code>logical(1)</code> Show progress. Not recommended if <code>BP = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>BP</th>
      <td><p><code>logical(1)</code> Allows the use of <span class="pkg">BiocParallel</span> within
<code>compare_motifs()</code>. See <code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/register'>BiocParallel::register()</a></code> to change the default
backend. Setting <code>BP = TRUE</code> is only recommended for comparing large numbers
of motifs (&gt;10,000). Furthermore, the behaviour of <code>progress = TRUE</code> is
changed if <code>BP = TRUE</code>; the default <span class="pkg">BiocParallel</span> progress bar will
be shown (which unfortunately is much less informative).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>matrix</code> if <code>compare.to</code> is missing; <code>data.frame</code> otherwise.</p><ul>
<li><p>PCC: 0 represents complete distance, &gt;0 similarity.</p></li>
<li><p>MPCC: 0 represents complete distance, 1 complete similarity.</p></li>
<li><p>EUCL: 0 represents complete similarity, &gt;0 distance.</p></li>
<li><p>MEUCL: 0 represents complete similarity, sqrt(2) complete distance.</p></li>
<li><p>SW: 0 represents complete distance, &gt;0 similarity.</p></li>
<li><p>MSW: 0 represents complete distance, 2 complete similarity.</p></li>
<li><p>KL: 0 represents complete similarity, &gt;0 distance.</p></li>
<li><p>MKL: 0 represents complete similarity, 4.62 complete distance.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Comparisons are calculated between two motifs at a time. All possible alignments
are scored, and the best score is reported. Scores are calculated per position
and summed, unless the 'mean' version of the specific metric is chosen. If using
a similarity metric, then the sum of scores will favour comparisons between
longer motifs; and for distance metrics, the sum of scores will favour
comparisons between short motifs. This can be avoided by using the 'mean' of
scores.</p><ul>
<li><p>PCC: Pearson correlation coefficient</p>
<p>Per position:</p>
<p><code>PCC = sum(col1 * col2) / sqrt(sum(col1^2) * sum(col2^2))</code></p></li>
<li><p>MPCC: Mean PCC</p>
<p><code>MPCC = mean(PCC)</code></p></li>
<li><p>EUCL: Euclidian distance</p>
<p>Per position:</p>
<p><code>EUCL = sqrt(sum((col1 - col2)^2)) / sqrt(2)</code></p></li>
<li><p>MEUCL: Mean EUCL</p>
<p><code>MEUCL = sum(EUCL) / ncol(alignment)</code></p></li>
<li><p>SW: Sandelin-Wasserman similarity</p>
<p>Per position:</p>
<p><code>SW = 2 - sum((col1 - col2)^2)</code></p></li>
<li><p>MSW: Mean SW</p>
<p><code>MSW = mean(SW)</code></p></li>
<li><p>KL: Kullback-Leibler divergence</p>
<p>Per position:</p>
<p><code>KL = 0.5 * sum(col1 * log(col1/col2) + col2 * log(col2/col1))</code></p></li>
<li><p>MKL: Mean Kullback-Leibler divergence</p>
<p><code>MKL = mean(KL)</code></p></li>
</ul>
    <p>To note regarding p-values: p-values are pre-computed using the
<code>make_DBscores</code> function. If not given, then uses a set of internal
precomputed p-values from the JASPAR2018 CORE motifs. These precalculated
scores are dependent on the length of the motifs being compared; this takes
into account that comparing small motifs with larger motifs leads to higher
scores, since the probability of finding a higher scoring alignment is
higher.</p>
<p>The default p-values have been precalculated for regular DNA motifs; they
are of little use for motifs with a different number of alphabet letters
(or even the <code>multifreq</code> slot).</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Choi I, Kwon J, Kim S (2004).
&#8220;Local feature frequency profile: a method to measure structural similarity in proteins.&#8221;
<em>PNAS</em>, <b>101</b>, 3797--3802.</p>
<p>Khan A, Fornes O, Stigliani A, Gheorghe M, Castro-Mondragon J, van der
Lee R, Bessy A, Cheneby J, Kulkarni SR, Tan G, Baranasic D, Arenillas D, Sandelin A, Vandepoele K, Lenhard B, Ballester B, Wasserman W, Parcy F, Mathelier A (2018).
&#8220;JASPAR 2018: update of the open-access database of transcription factor binding profiles and its web framework.&#8221;
<em>Nucleic Acids Research</em>, <b>46</b>, D260-D266.</p>
<p>Pietrokovski S (1996).
&#8220;Searching databases of conserved sequence regions by aligning protein multiple-alignments.&#8221;
<em>Nucleic Acids Research</em>, <b>24</b>, 3836--3845.</p>
<p>Roepcke S, Grossmann S, Rahmann S, Vingron M (2005).
&#8220;T-Reg Comparator: an analysis tool for the comparison of position weight matrices.&#8221;
<em>Nucleic Acids Research</em>, <b>33</b>, W438--W441.</p>
<p>Sandelin A, Wasserman W (2004).
&#8220;Constrained binding site diversity within families of transcription factors enhances pattern discovery bioinformatics.&#8221;
<em>Journal of Molecular Biology</em>, <b>338</b>(2), 207--215.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='convert_motifs.html'>convert_motifs()</a></code>, <code><a href='motif_tree.html'>motif_tree()</a></code>, <code><a href='view_motifs.html'>view_motifs()</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>motif1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_motif.html'>create_motif</a></span>()
<span class='no'>motif2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_motif.html'>create_motif</a></span>()
<span class='no'>motif1vs2</span> <span class='kw'>&lt;-</span> <span class='fu'>compare_motifs</span>(<span class='fu'>list</span>(<span class='no'>motif1</span>, <span class='no'>motif2</span>), <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"MPCC"</span>)</div><div class='output co'>#&gt;    0%100%</div><div class='input'><span class='co'>## to get a dist object:</span>
<span class='fu'>as.dist</span>(<span class='fl'>1</span> - <span class='no'>motif1vs2</span>)</div><div class='output co'>#&gt;                          motif
#&gt; motif [duplicated #1] 0.476604</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Benjamin Tremblay, <a href='mailto:b2tremblay@uwaterloo.ca'>b2tremblay@uwaterloo.ca</a>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Benjamin Tremblay.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

