<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert <a href='universalmotif-class.html'>universalmotif</a> type. — convert_type • universalmotif</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Convert <a href='universalmotif-class.html'>universalmotif</a> type. — convert_type" />

<meta property="og:description" content="Switch between position count matrix (PCM), position probability matrix
(PPM), position weight matrix (PWM), and information count matrix (ICM)
types." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">universalmotif</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/AdvancedUsage.html">Advanced motif and sequence manipulation</a>
    </li>
    <li>
      <a href="../articles/Introduction.html">Introduction to the "universalmotif" package</a>
    </li>
    <li>
      <a href="../articles/IntroductionToSequenceMotifs.html">Introduction to sequence motifs</a>
    </li>
    <li>
      <a href="../articles/MotifManipulation.html">Motif import, export, visualization, and manipulation</a>
    </li>
    <li>
      <a href="../articles/SequenceSearches.html">Sequence creation, manipulation, and scanning</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bjmt/universalmotif">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert <a href='universalmotif-class.html'>universalmotif</a> type.</h1>
    <small class="dont-index">Source: <a href='https://github.com/bjmt/universalmotif/blob/master/R/convert_type.R'><code>R/convert_type.R</code></a></small>
    <div class="hidden name"><code>convert_type.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Switch between position count matrix (PCM), position probability matrix
(PPM), position weight matrix (PWM), and information count matrix (ICM)
types.</p>
    
    </div>

    <pre class="usage"><span class='fu'>convert_type</span>(<span class='no'>motifs</span>, <span class='no'>type</span>, <span class='no'>pseudocount</span>, <span class='kw'>nsize_correction</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>relative_entropy</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>motifs</th>
      <td><p>See <code><a href='convert_motifs.html'>convert_motifs()</a></code> for acceptable formats.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p><code>character(1)</code> One of <code>c('PCM', 'PPM', 'PWM', 'ICM')</code>.</p></td>
    </tr>
    <tr>
      <th>pseudocount</th>
      <td><p><code>numeric(1)</code> Correction to be applied to prevent <code>-Inf</code>
from apearing in PWM matrices. If missing, the pseudocount stored in the
<a href='universalmotif-class.html'>universalmotif</a> 'pseudocount' slot will be
used, which defaults to 0.8; the suggested value from
Nishida et al. (2009)
.</p></td>
    </tr>
    <tr>
      <th>nsize_correction</th>
      <td><p><code>logical(1)</code> If true, the ICM
at each position will be corrected to account
for small sample sizes. Only used if <code>relative_entropy = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>relative_entropy</th>
      <td><p><code>logical(1)</code> If true, the ICM will be
calculated as relative entropy. See details.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>See <code><a href='convert_motifs.html'>convert_motifs()</a></code> for possible output motif objects.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Position count matrix (PCM), also known as position frequency matrix
(PFM). For n sequences from which the motif was built, each position is
represented by the numbers of each letter at that position. In theory
all positions should have sums equal to n, but not all databases are
this consistent. If converting from another type to PCM, column sums
will be equal to the 'nsites' slot; if empty, 100 is used.</p>
<p>Position probability matrix (PPM), also known as position frequency
matrix (PFM). At each position, the probability of individual letters
is calculated by dividing the count for that letter by the total sum of
counts at that position (<code>letter_count / position_total</code>).
As a result, each position will sum to 1. Letters with counts of 0 will
thus have a probability of 0, which can be undesirable when searching for
motifs in a set of sequences. To avoid this a pseudocount can be added
(<code>(letter_count + pseudocount) / (position_total + pseudocount)</code>).</p>
<p>Position weight matrix (PWM; Stormo et al. (1982)
),
also known as position-specific weight
matrix (PSWM), position-specific scoring matrix (PSSM), or
log-odds matrix. At each position, each letter is represented by it's
log-likelihood (<code>log2(letter_probability / background_probility)</code>),
which is normalized using the background letter frequencies. A PWM matrix
is constructed from a PPM; if any position has 0-probability letters to
which pseudocounts were not added, then the final log-likelihodd of these
letters will be <code>-Inf</code>.</p>
<p>Information content matrix (ICM; Schneider and Stephens (1990)
).
An ICM is a PPM where each letter probability is multiplied by the total
information content at that position. The information content of each
position is determined as: <code>totalIC - Hi</code>, where the total information
totalIC</p>
<p><code>totalIC &lt;- log2(alphabet_length)</code>, and the Shannon entropy
(Shannon 1948)
 for a specific
position (Hi)</p>
<p><code>Hi &lt;- -sum(sapply(alphabet_frequencies, function(x) x * log(2))</code>.</p>
<p>As a result, the total sum or height of each position is representative of
it's sequence conservation, measured in the unit 'bits', which is a unit
of energy (Schneider (1991)
; see
<a href='https://fr-s-schneider.ncifcrf.gov/logorecommendations.html'>https://fr-s-schneider.ncifcrf.gov/logorecommendations.html</a>
for more information). However not all programs will calculate
information content the same; some will 'correct' the total information
content at each position using a correction factor as described by
Schneider et al. (1986)
. This correction can
applied by setting <code>nsize_correction = TRUE</code>, however it will only
be applied if the 'nsites' slot is not empty. This is done using
<code>TFBSTools:::schneider_correction</code>
(Tan and Lenhard 2016)
. As such, converting from an ICM to
which some form of correction has been applied will result in a
PCM/PPM/PWM with slight inaccuracies.</p>
<p>Another method of calculating information content is calculating the
relative entropy, also known as Kullback-Leibler divergence
(Kullback and Leibler 1951)
. This accounts for background
frequencies, which
can be useful for genomes with a heavy imbalance in letter frequencies.
For each position, the individual letter frequencies are calculated as
<code>letter_freq * log2(letter_freq / bkg_freq)</code>. When calculating
information content using Shannon entropy, the maximum content for
each position will always be <code>log2(alphabet_length)</code>; this does
not hold for information content calculated as relative entropy.
Please note that conversion from ICM assumes the information content
was <em>not</em> calculated as relative entropy.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Kullback S, Leibler RA (1951).
&#8220;On information and sufficiency.&#8221;
<em>The Annals of Mathematical Statistics</em>, <b>22</b>, 79-86.</p>
<p>Nishida K, Frith MC, Nakai K (2009).
&#8220;Pseudocounts for transcription factor binding sites.&#8221;
<em>Nucleic Acids Research</em>, <b>37</b>, 939-944.</p>
<p>Schneider TD, Stormo GD, Gold L, Ehrenfeucht A (1986).
&#8220;Information content of binding sites on nucleotide sequences.&#8221;
<em>Journal of Molecular Biology</em>, <b>188</b>, 415-431.</p>
<p>Schneider TD, Stephens RM (1990).
&#8220;Sequence Logos: A New Way to Display Consensus Sequences.&#8221;
<em>Nucleic Acids Research</em>, <b>18</b>, 6097-6100.</p>
<p>Schneider TD (1991).
&#8220;Theory of Molecular Machines. II. Energy Dissipation from Molecular Machines.&#8221;
<em>Journal of Theoretical Biology</em>, <b>148</b>, 125-137.</p>
<p>Shannon CE (1948).
&#8220;A Mathematical Theory of Communication.&#8221;
<em>Bell System Technical Journal</em>, <b>27</b>, 379-423.</p>
<p>Stormo GD, Schneider TD, Gold L, Ehrenfeucht A (1982).
&#8220;Use of the Perceptron algorithm to distinguish translational initiation sites in E. coli.&#8221;
<em>Nucleic Acids Research</em>, <b>10</b>, 2997-3011.</p>
<p>Tan G, Lenhard B (2016).
&#8220;TFBSTools: an R/Bioconductor package for transcription factor binding site analysis.&#8221;
<em>Bioinformatics</em>, <b>32</b>, 1555-1556.
doi: <a href='http://doi.org/10.1093/bioinformatics/btw024'>10.1093/bioinformatics/btw024</a>
, <a href='http://bioinformatics.oxfordjournals.org/content/32/10/1555'>http://bioinformatics.oxfordjournals.org/content/32/10/1555</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='convert_motifs.html'>convert_motifs()</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>jaspar.pcm</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='read_jaspar.html'>read_jaspar</a></span>(<span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='st'>"jaspar.txt"</span>,
                                      <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"universalmotif"</span>))

<span class='co'>## pseudocounts default to 0.8</span>
<span class='no'>jaspar.pwm</span> <span class='kw'>&lt;-</span> <span class='fu'>convert_type</span>(<span class='no'>jaspar.pcm</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"PPM"</span>)

<span class='co'>## setting pseudocounts to 0 will prevent any correction from being</span>
<span class='co'>## applied to PPM/PWM matrices</span>
<span class='no'>jaspar.pwm</span> <span class='kw'>&lt;-</span> <span class='fu'>convert_type</span>(<span class='no'>jaspar.pcm</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"PWM"</span>, <span class='kw'>pseudocount</span> <span class='kw'>=</span> <span class='fl'>0</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Benjamin Tremblay, <a href='mailto:b2tremblay@uwaterloo.ca'>b2tremblay@uwaterloo.ca</a>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Benjamin Tremblay.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

